---
title: "PCBP2 Comparisons"
author: "Justin Bogias"
date: "2024-09-10"
output:
  rmdformats::readthedown
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

```{r}
library(rmdformats)
library(dplyr)
library(readr)
library(tidyr)
library(stringr)
library(ggplot2)
library(ggbeeswarm)
library(magrittr)
library(tibble)
library(pheatmap)
library(cowplot)
library(here)

source(here("R/get_dtelist.R"))
source(here("R/de_testing.R"))
```

The code below details the differential expression and usage in each method

```{r}
gene_txp_anno <- read_csv(here("data/annotations/grch38_103_df.csv.gz"))

txp_gene_ensdb_lengths <- read_csv(
  here("data/annotations/txp_gene_ensdb_lengths.csv.gz")
  )

transcript_key <- dplyr::select(gene_txp_anno,
                                c("transcript_id",
                                  "transcript_name"))

tx_anno <- dplyr::mutate(txp_gene_ensdb_lengths,
                         transcript_id = tx_id) %>%
  as.data.frame() %>%
  column_to_rownames("tx_id") %>%
  left_join(transcript_key, by = "transcript_id")
```

```{r}
hisat2_counts <- read_rds(here("data/counts/salmon_object_hisat2_grch38.rds"))

hisat2_dtelist <- get_dtelist(hisat2_counts, annotations = tx_anno,
                              normalise = TRUE)

star_counts <- read_rds(here("data/counts/salmon_object_star_grch38.rds"))

star_dtelist <- get_dtelist(star_counts, annotations = tx_anno)

salmon_counts <- read_rds(here("data/counts/SA-SalmonObject_HTR8_grch38.rds"))

salmon_dtelist <- get_dtelist(salmon_counts, annotations = tx_anno)

kallisto_counts <- read_rds(
  here("data/counts/kallisto_object_HTR8_grch38_global.rds")
  )

kallisto_dtelist <- get_dtelist(kallisto_counts, annotations = tx_anno)
```

# Get Multi-Mapping Data
## Quant Files
```{r}
hisat2_quant_files <- character()
for(i in list.files(here("data/counts/HISAT2/hisat2_quant"))) {
 quant <- paste0(here("data/counts/HISAT2/hisat2_quant/"), i, "/quant.sf")
 hisat2_quant_files <- c(hisat2_quant_files, quant)
}
  
sa_quant_files <- character()
for(i in list.files(here("data/counts/SA/quant_files"))) {
 quant <- paste0(here("data/counts/SA/quant_files/"), i)
 sa_quant_files <- c(sa_quant_files, quant)
}

star_quant_files <- character()
for(i in list.files(here("data/counts/STAR/quant_files"))) {
 quant <- paste0(here("data/counts/STAR/quant_files/"), i)
 star_quant_files <- c(star_quant_files, quant)
}
```

## Ambiguous Read Files
```{r}
hisat2_ambig_files <- character()
for(i in list.files(here("data/counts/HISAT2/hisat2_quant"))) {
  ambig <- paste0(here("data/counts/HISAT2/hisat2_quant/"), i,
                  "/aux_info/ambig_info.tsv")
  hisat2_ambig_files <- c(hisat2_ambig_files, ambig)
}

sa_ambig_files <- character()
for(i in list.files(here("data/counts/SA/ambig_info"))) {
  ambig <- paste0(here("data/counts/SA/ambig_info/"), i)
  sa_ambig_files <- c(sa_ambig_files, ambig)
}

star_ambig_files <- character()
for(i in list.files(here("data/counts/STAR/ambig_info"))) {
  ambig <- paste0(here("data/counts/STAR/ambig_info/"), i)
  star_ambig_files <- c(star_ambig_files, ambig)
}
```

## Set multi-mapping info function
```{r}
multi_mapping <- function(quant, ambig) {
  x <- cbind(quant, ambig) %>%
    mutate(sum = TPM + NumReads + UniqueCount + AmbigCount) %>%
    dplyr::filter(sum > 0) %>%
    dplyr::select(-sum) %>%
    mutate(multimap = NumReads-UniqueCount) %>%
    mutate(percent_est_from_multimap = multimap/NumReads,
           percent_est_from_multimap = replace_na(percent_est_from_multimap, 0),
           percent_ambigs_used = multimap/AmbigCount,
           percent_ambigs_used = replace_na(percent_ambigs_used, 0),
           status = case_when(
             UniqueCount == NumReads & UniqueCount > 0 ~ "all_unique_reads",
             UniqueCount == 0 & NumReads == 0 & AmbigCount > 0 ~ "all_ambig_no_counts",
             multimap == NumReads ~ "all_ambig_reads",
             AmbigCount != NumReads & UniqueCount != NumReads & percent_est_from_multimap > 0.5 ~ "more_multi",
             AmbigCount != NumReads & UniqueCount != NumReads & percent_est_from_multimap < 0.5 ~ "more_unique",
             percent_est_from_multimap == 0.5 ~ "equal"
           ),
           status = factor(status, levels = c("all_unique_reads",
                                              "more_unique",
                                              "equal",
                                              "more_multi",
                                              "all_ambig_reads")))
  
  return(x)
}
```

## Get Multi-mapping for each Aligner
### HISAT2
```{r}
hisat2_quants <- list()
for(i in 1:length(hisat2_quant_files)) {
  hisat2_quants[[i]] <- read_delim(
    hisat2_quant_files[[i]]
  )
}
names(hisat2_quants) <- list.files(here("data/counts/HISAT2/hisat2_quant"))

hisat2_ambigs <- list()
for(i in 1:length(hisat2_ambig_files)) {
  hisat2_ambigs[[i]] <- read_tsv(
    hisat2_ambig_files[[i]]
  )
}
names(hisat2_ambigs) <- list.files(here("data/counts/HISAT2/hisat2_quant"))


hisat2_multimap <- list()
for(i in 1:length(hisat2_quants)) {
  hisat2_multimap[[i]] <- multi_mapping(quant = hisat2_quants[[i]], 
                                        ambig = hisat2_ambigs[[i]])
}
names(hisat2_multimap) <- names(hisat2_quants)
```

### Selective Alignment
```{r}
sa_quants <- list()
for(i in 1:length(sa_quant_files)) {
  sa_quants[[i]] <- read_delim(
    sa_quant_files[[i]]
  )
}
names(sa_quants) <- basename(sa_quant_files) %>%
  str_remove("quant_") %>%
  str_remove(".sf")

sa_ambigs <- list()
for(i in 1:length(sa_ambig_files)) {
  sa_ambigs[[i]] <- read_tsv(
    sa_ambig_files[[i]]
  )
}
names(sa_ambigs) <- basename(sa_ambig_files) %>%
  str_remove("_ambig_info") %>%
  str_remove(".tsv")

sa_multimap <- list()
for(i in 1:length(sa_quants)) {
  sa_multimap[[i]] <- multi_mapping(quant = sa_quants[[i]], 
                                    ambig = sa_ambigs[[i]])
}
names(sa_multimap) <- names(sa_quants)
```

### STAR
```{r}
star_quants <- list()
for(i in 1:length(star_quant_files)) {
  star_quants[[i]] <- read_delim(
    star_quant_files[[i]]
  )
}
names(star_quants) <- basename(star_quant_files) %>%
  str_remove("quant_") %>%
  str_remove(".sf")

star_ambigs <- list()
for(i in 1:length(star_ambig_files)) {
  star_ambigs[[i]] <- read_tsv(
    star_ambig_files[[i]]
  )
}
names(star_ambigs) <- basename(star_ambig_files) %>%
  str_remove("_ambig_info") %>%
  str_remove(".tsv")

star_multimap <- list()
for(i in 1:length(star_quants)) {
  star_multimap[[i]] <- multi_mapping(quant = star_quants[[i]], 
                                      ambig = star_ambigs[[i]])
}
names(star_multimap) <- names(star_quants)
```

# Heatmaps
Check to see how many of the total identified *PCBP2* are found in each method
```{r}
pcbp2_anno <- txp_gene_ensdb_lengths %>%
  dplyr::filter(gene_id == "ENSG00000197111")

pcbp2 <- transcript_key %>%
  dplyr::filter(transcript_id %in% pcbp2_anno$tx_id)

hisat2_df <- hisat2_counts$counts %>%
  as.data.frame() %>%
  set_colnames(paste0("HISAT2_", basename(colnames(hisat2_counts$counts)))) %>%
  rownames_to_column("transcript_id") %>%
  dplyr::mutate(transcript_id = str_remove(transcript_id, "\\..*")) %>%
  dplyr::filter(transcript_id %in% pcbp2$transcript_id)

hisat2_pcbp2 <- hisat2_df %>%
  pivot_longer(cols = starts_with("HISAT2"),
               names_to = "ID",
               values_to = "counts") %>%
  dplyr::mutate(group = "hisat2")

hisat2_pheat <- hisat2_df %>% 
  as.data.frame() %>%
  column_to_rownames("transcript_id") %>%
  rowMeans() %>% 
  as.data.frame() %>%
  set_colnames("hisat2_counts") %>% 
  dplyr::arrange(desc(hisat2_counts)) %>%
  dplyr::mutate(hisat2_counts = log2(hisat2_counts+1)) %>%
  rownames_to_column("transcript_id") %>%
  left_join(pcbp2, by = "transcript_id") %>%
  dplyr::select(-"transcript_id")
# 26 of 26 found

kallisto_df <- kallisto_counts$counts %>%
  as.data.frame() %>%
  set_colnames(paste0("Kallisto_", 
                      basename(colnames(kallisto_counts$counts)))) %>%
  rownames_to_column("transcript_id") %>%
  dplyr::mutate(transcript_id = str_remove(transcript_id, "\\..*")) %>%
  dplyr::filter(transcript_id %in% pcbp2$transcript_id)

kallisto_pcbp2 <- kallisto_df %>%
  pivot_longer(cols = starts_with("Kallisto"),
               names_to = "ID",
               values_to = "counts") %>%
  dplyr::mutate(group = "kallisto")

kallisto_pheat <- kallisto_df %>% 
  as.data.frame() %>%
  column_to_rownames("transcript_id") %>%
  rowMeans() %>% 
  as.data.frame() %>%
  set_colnames("kallisto_counts") %>% 
  dplyr::arrange(desc(kallisto_counts)) %>%
  dplyr::mutate(kallisto_counts = log2(kallisto_counts+1)) %>%
  rownames_to_column("transcript_id") %>%
  left_join(pcbp2, by = "transcript_id") %>%
  dplyr::select(-"transcript_id")
# 24 of 26 found

star_df <- star_counts$counts %>%
  as.data.frame() %>%
  set_colnames(paste0("STAR_", basename(colnames(star_counts$counts)))) %>%
  rownames_to_column("transcript_id") %>%
  dplyr::mutate(transcript_id = str_remove(transcript_id, "\\..*")) %>%
  dplyr::filter(transcript_id %in% pcbp2$transcript_id) 

star_pcbp2 <- star_df %>%
  pivot_longer(cols = starts_with("STAR"),
               names_to = "ID",
               values_to = "counts") %>%
  dplyr::mutate(group = "star")

star_pheat <- star_df %>% 
  as.data.frame() %>%
  column_to_rownames("transcript_id") %>%
  rowMeans() %>% 
  as.data.frame() %>%
  set_colnames("star_counts") %>% 
  dplyr::arrange(desc(star_counts)) %>%
  dplyr::mutate(star_counts = log2(star_counts+1)) %>%
  rownames_to_column("transcript_id") %>%
  left_join(pcbp2, by = "transcript_id") %>%
  dplyr::select(-"transcript_id")
# 25 of 26 found

salmon_df <- salmon_counts$counts %>%
  as.data.frame() %>%
  set_colnames(paste0("Salmon_", basename(colnames(salmon_counts$counts)))) %>%
  rownames_to_column("transcript_id") %>%
  dplyr::mutate(transcript_id = str_remove(transcript_id, "\\..*")) %>%
  dplyr::filter(transcript_id %in% pcbp2$transcript_id)

salmon_pcbp2 <- salmon_df %>%
  pivot_longer(cols = starts_with("Salmon"),
               names_to = "ID",
               values_to = "counts") %>%
  dplyr::mutate(group = "salmon")

salmon_pheat <- salmon_df %>% 
  as.data.frame() %>%
  column_to_rownames("transcript_id") %>%
  rowMeans() %>% 
  as.data.frame() %>%
  set_colnames("salmon_counts") %>% 
  dplyr::arrange(desc(salmon_counts)) %>%
  dplyr::mutate(salmon_counts = log2(salmon_counts+1)) %>%
  rownames_to_column("transcript_id") %>%
  left_join(pcbp2, by = "transcript_id") %>%
  dplyr::select(-"transcript_id")
# 25 of 26 found

pcbp2_pheat <- left_join(hisat2_df, kallisto_df,
                         by = "transcript_id") %>%
  left_join(salmon_df, by = "transcript_id") %>%
  left_join(star_df, by = "transcript_id") %>%
  left_join(pcbp2, by = "transcript_id") %>%
  column_to_rownames("transcript_name") %>%
  dplyr::select(-"transcript_id")

pheatmap(log2(pcbp2_pheat+1),
         border_color = NA)

pcbp2_df <- rbind(hisat2_pcbp2, kallisto_pcbp2, star_pcbp2, salmon_pcbp2) %>%
  left_join(hisat2_dtelist$samples %>% 
              dplyr::select("exp_group" = group, ID),
            by = "ID") %>%
  left_join(transcript_key,
            by = "transcript_id") %>%
  mutate(counts = counts + 0.00001)
```

## New heatmap
```{r}
column_order <- c("HISAT2_SRR13401120", "HISAT2_SRR13401121",
                  "HISAT2_SRR13401122", "HISAT2_SRR13401123",
                  "Kallisto_SRR13401120", "Kallisto_SRR13401121",
                  "Kallisto_SRR13401122", "Kallisto_SRR13401123",
                  "Salmon_SRR13401120", "Salmon_SRR13401121",
                  "Salmon_SRR13401122", "Salmon_SRR13401123",
                  "STAR_SRR13401120", "STAR_SRR13401121",
                  "STAR_SRR13401122", "STAR_SRR13401123",
                  "HISAT2_SRR13401116", "HISAT2_SRR13401117",
                  "HISAT2_SRR13401118", "HISAT2_SRR13401119",
                  "Kallisto_SRR13401116", "Kallisto_SRR13401117",
                  "Kallisto_SRR13401118", "Kallisto_SRR13401119",
                  "Salmon_SRR13401116", "Salmon_SRR13401117",
                  "Salmon_SRR13401118", "Salmon_SRR13401119",
                  "STAR_SRR13401116", "STAR_SRR13401117",
                  "STAR_SRR13401118", "STAR_SRR13401119")

col_anno <- rep(c("Control", "Knockout"), c(16, 16)) %>%
  as.data.frame() %>%
  set_rownames(column_order) %>%
  set_colnames("Group") %>%
  mutate(Method = rep(
    c(
      rep("HISAT2", 4),
      rep("Kallisto", 4), 
      rep("Salmon", 4),
      rep("STAR", 4)
      ), 2))

hisat2_anno <- pcbp2_pheat %>%
  dplyr::select(starts_with("HISAT2")) %>%
  mutate(HISAT2 = case_when(
    rowSums(.) == 0 ~ "FALSE",
    .default = "TRUE"
  )) %>%
  rownames_to_column("transcript_name") %>%
  dplyr::select("transcript_name", "HISAT2")

kallisto_anno <- pcbp2_pheat %>%
  dplyr::select(starts_with("Kallisto")) %>%
  mutate(Kallisto = case_when(
    rowSums(.) == 0 ~ "FALSE",
    .default = "TRUE"
  )) %>%
  rownames_to_column("transcript_name") %>%
  dplyr::select("transcript_name", "Kallisto")

row_anno <- full_join(hisat2_anno, kallisto_anno, by = "transcript_name")

salmon_anno <- pcbp2_pheat %>%
  dplyr::select(starts_with("Salmon")) %>%
  mutate(Salmon = case_when(
    rowSums(.) == 0 ~ "FALSE",
    .default = "TRUE"
  )) %>%
  rownames_to_column("transcript_name") %>%
  dplyr::select("transcript_name", "Salmon")

row_anno <- full_join(row_anno, salmon_anno, by = "transcript_name")

star_anno <- pcbp2_pheat %>%
  dplyr::select(starts_with("STAR")) %>%
  mutate(STAR = case_when(
    rowSums(.) == 0 ~ "FALSE",
    .default = "TRUE"
  )) %>%
  rownames_to_column("transcript_name") %>%
  dplyr::select("transcript_name", "STAR")

row_anno <- full_join(row_anno, star_anno, by = "transcript_name")

row_anno <- row_anno %>%
  column_to_rownames("transcript_name")

ann_colours <- list(Group = c(Control = "darkblue",
                              Knockout = "orange"),
                    Method = c(HISAT2 = "black",
                               Kallisto = "brown",
                               Salmon = "salmon",
                               STAR = "pink"),
                    HISAT2 = c(`TRUE` = "green",
                                 `FALSE` = "red"),
                    Kallisto = c(`TRUE` = "green",
                                 `FALSE` = "red"),
                    Salmon = c(`TRUE` = "green",
                                 `FALSE` = "red"),
                    STAR = c(`TRUE` = "green",
                                 `FALSE` = "red")
                    )

log2(pcbp2_pheat+1)[, column_order] %>%
  pheatmap(border_color = NA,
           cluster_rows = TRUE,
           cluster_cols = FALSE,
           gaps_col = 16,
           annotation_col = col_anno,
           annotation_row = row_anno,
           annotation_colors = ann_colours,
           show_colnames = FALSE)

# Use grid.ls(grid.force()) to break the heatmap up into grid elements
# Then steadily turn each element white as you go along until you get the desired layout, using grid.gedit("GRID.text.13423", gp = gpar(col = "white"))
```

# Plot *PCBP2* Counts
```{r}
tx_ids <- pcbp2_df$transcript_id %>% unique()

for(tx in tx_ids) {
  tx_name <- pcbp2_df %>%
    dplyr::filter(transcript_id == tx) %>%
    dplyr::select(transcript_name) %>%
    unique() %>%
    as.character()
  
  pcbp2_plot <- pcbp2_df %>%
    #dplyr::mutate("line_dat" = rep(1:8, 104)) %>%
    dplyr::filter(transcript_id == tx) %>%
    mutate(group = str_to_title(group),
           group = case_when(group == "Star" ~ "STAR",
                             group == "Hisat2" ~ "HISAT2",
                             .default = group),
           exp_group = str_to_title(exp_group)) %>%
    ggplot(aes(x = exp_group, y = counts, colour = exp_group)) +
    #geom_boxplot(colour = "black", outlier.shape = NA) +
    geom_quasirandom(width = 0.2, size = 2.5) +
    geom_line() +
    #scale_fill_manual(values = c("royalblue", "darkorange")) +
    scale_colour_manual(values = c("royalblue", "darkorange")) +
    #guides(colour = "none") +
    facet_grid(~ group) +
    labs(x = "",
         y = "Transcript Expression",
         colour = "Experimental\nGroup") +
    ggtitle(tx_name) +
    theme_bw() +
    theme(axis.title = element_text(colour = "black", size = 14),
          axis.text.y = element_text(colour = "black", size = 12),
          axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          strip.text = element_text(size = 14, colour = "black"),
          strip.background = element_rect(fill = "lightblue"),
          legend.title = element_text(colour = "black", size = 14,
                                      face = "bold"),
          legend.text = element_text(colour = "black", size = 12))
  
  ggsave(plot = pcbp2_plot,
         filename = paste0(tx_name, "_counts.png"),
         path = here("figures/pcbp2_plots/counts/"),
         device = "png",
         height = 125,
         width = 200,
         units = "mm",
         dpi = 400,
         create.dir = TRUE)
}
```

# Multi-Mapping Counts by Experimental Group
```{r}
big_multi_df <- list()
big_multi_df[["hisat2"]] <- hisat2_multimap
big_multi_df[["star"]] <- star_multimap
big_multi_df[["salmon"]] <- sa_multimap

multimap_df <- data.frame()

for(i in names(big_multi_df)) {
  for(j in names(hisat2_multimap)) {
    x <- big_multi_df[[i]][[j]] %>%
      dplyr::mutate(transcript_id = str_remove(Name, "\\..*")) %>%
      dplyr::filter(transcript_id %in% pcbp2$transcript_id) %>%
      dplyr::mutate("ID" = j,
                    "group" = i)
    
    multimap_df <- rbind(multimap_df, x)
  }
}
```

## Without Kallisto Counts
We couldn't get any multi-mapping counts from *Kallisto* since it does not share any with us.
```{r}
multimap_df <- multimap_df %>%
  left_join(hisat2_dtelist$samples %>%
              dplyr::select("exp_group" = group, ID),
            by = "ID") %>%
  left_join(transcript_key,
            by = "transcript_id") %>%
  dplyr::mutate(exp_group = str_to_title(exp_group),
                group = str_to_title(group),
                group = case_when(group == "Star" ~ "STAR",
                                  group == "Hisat2" ~ "HISAT2",
                                  .default = group))

for(tx in unique(multimap_df$transcript_id)) {
  tx_name <- multimap_df %>%
    dplyr::filter(transcript_id == tx) %>%
    dplyr::select(transcript_name) %>%
    unique() %>%
    as.character()
  
  pcbp2_multimapping <- multimap_df %>%
    dplyr::filter(transcript_id == tx) %>%
    ggplot(aes(x = exp_group, y = multimap, colour = exp_group)) +
    #geom_boxplot(colour = "black", outlier.shape = NA) +
    geom_quasirandom(width = 0.2, size = 2.5) +
    geom_line() +
    #scale_fill_manual(values = c("royalblue", "darkorange")) +
    scale_colour_manual(values = c("royalblue", "darkorange")) +
    facet_wrap(~ group) +
    ggtitle(tx_name) +
    labs(x = "",
         y = "Multimapped Counts",
         colour = "Experimental\nGroup") +
    theme_bw() +
    theme(axis.title = element_text(colour = "black", size = 14),
          axis.text.y = element_text(colour = "black", size = 12),
          axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          strip.text = element_text(size = 14, colour = "black"),
          strip.background = element_rect(fill = "lightblue"),
          legend.title = element_text(colour = "black", size = 14,
                                      face = "bold"),
          legend.text = element_text(colour = "black", size = 12))
  
    ggsave(plot = pcbp2_multimapping,
         filename = paste0(tx_name, "_multimapping.png"),
         path = here("figures/pcbp2_plots/multimapping/"),
         device = "png",
         height = 125,
         width = 200,
         units = "mm",
         dpi = 400,
         create.dir = TRUE)
}
```

## Create With Kallisto Counts
Here we just plot *Kallisto* with empty counts, its just a visual thing
```{r}
tx_ids <- pcbp2_df$transcript_id %>% unique()

pcbp2_df_nokal <- pcbp2_df %>%
  dplyr::filter(!group == "kallisto")

for(tx in tx_ids) {
  tx_name <- pcbp2_df %>%
    dplyr::filter(transcript_id == tx) %>%
    dplyr::select(transcript_name) %>%
    unique() %>%
    as.character()
  
  pcbp2_plot <- pcbp2_df %>%
    dplyr::filter(transcript_id == tx) %>%
    mutate(group = str_to_title(group),
           group = case_when(group == "Star" ~ "STAR",
                             group == "Hisat2" ~ "HISAT2",
                             .default = group)) %>%
    ggplot(aes(x = exp_group,
               y = cpm(counts, log = TRUE), 
               colour = exp_group)) +
    #geom_boxplot(colour = "black", outlier.shape = NA) +
    geom_quasirandom(size = 3) +
    scale_fill_manual(values = c("royalblue", "darkorange")) +
    scale_colour_manual(values = c("royalblue", "darkorange")) +
    scale_y_continuous(limits = c(0, 17)) +
    facet_grid(~ group) +
    labs(x = "",
         y = "Transcript Expression (log2 CPM)",
         fill = "Experimental\nGroup") +
    ggtitle(tx_name) +
    theme_bw() +
    theme(axis.title = element_text(colour = "black", size = 14),
          axis.text.y = element_text(colour = "black", size = 12),
          axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          strip.text = element_text(size = 14, colour = "black"),
          strip.background = element_rect(fill = "lightblue"),
          plot.title = element_text(colour = "black", size = 18),
          legend.position = "none")
  # legend.title = element_text(colour = "black", size = 14,
  #                             face = "bold"),
  # legend.text = element_text(colour = "black", size = 12))
  
  pcbp2_multimapping <- multimap_df %>%
    dplyr::mutate(group = str_to_title(group),
                  group = case_when(group == "Star" ~ "STAR",
                                    group == "Hisat" ~ "HISAT2",
                                    .default = group)) %>%
    dplyr::filter(transcript_id == tx) %>%
    dplyr::select(transcript_id, percent_est_from_multimap,
                  exp_group, group) %>%
    rbind(c("ENST00000437231", -0.05, "Control", "Kallisto")) %>%
    mutate(
      percent_est_from_multimap = as.numeric(percent_est_from_multimap)
      ) %>%
    ggplot(aes(x = exp_group, y = 100*percent_est_from_multimap,
               colour = exp_group)) +
    geom_quasirandom(size = 3) +
    scale_y_continuous(limits = c(0, 100)) +
    scale_colour_manual(values = c("royalblue", "darkorange")) +
    facet_grid(~ group) +
    labs(x = "",
         y = "Multimapped Counts (% of total)",
         colour = "Experimental\nGroup") +
    theme_bw() +
    theme(axis.title = element_text(colour = "black", size = 14),
          axis.text.y = element_text(colour = "black", size = 12),
          axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          strip.text = element_text(size = 14, colour = "black"),
          strip.background = element_rect(fill = "lightblue"),
          legend.title = element_text(colour = "black", size = 14,
                                      face = "bold"),
          legend.text = element_text(colour = "black", size = 12),
          legend.position = "right")
  
  pcbp2_legend <- cowplot::get_legend(pcbp2_multimapping)
  
  pcbp2_cowplot <- cowplot::plot_grid(pcbp2_plot, 
                                      pcbp2_multimapping + 
                                        theme(legend.position = "none"),
                                      ncol = 1)
  
  pcbp2_cowplot_legend <- cowplot::plot_grid(pcbp2_cowplot,
                                             pcbp2_legend,
                                             nrow = 1,
                                             rel_widths = c(1, 0.13))
  
  ggsave(plot = pcbp2_cowplot_legend,
         filename = paste0(tx_name, "_cowplot.png"),
         path = here("figures/pcbp2_plots/cowplots_with_kallisto/"),
         device = "png",
         height = 225,
         width = 275,
         units = "mm",
         dpi = 400,
         create.dir = TRUE)
}

pcbp2_cowplot_legend
```

This plot needs to be made manually, the code is here, I just need to include the plotting
```{r eval=FALSE}
# I just changed some of the legend positioning stuff for this and ran it manually
fig7 <- cowplot::plot_grid(p202_plot,
                   p206_plot,
                   p209_plot,
                   pcbp2_legend,
                   ncol = 1,
                   labels = c("A)", "B)", "C)"),
                   rel_heights = c(1,1,1,0.05))

ggsave(plot = fig7,
       filename = "fig7_pcbp2_cowplot.png",
       path = here("figures/pcbp2_plots/"),
       device = "png",
       height = 500,
       width = 300,
       units = "mm",
       dpi = 400,
       create.dir = TRUE)
```

## facet plot
```{r}
tx_ids <- pcbp2_df$transcript_id %>% unique()

pcbp2_df_nokal <- pcbp2_df %>%
  dplyr::filter(!group == "kallisto")

for(tx in tx_ids) {
  tx_name <- pcbp2_df_nokal %>%
    dplyr::filter(transcript_id == tx) %>%
    dplyr::select(transcript_name) %>%
    unique() %>%
    as.character()
  
  pcbp2_plot <- pcbp2_df_nokal %>%
    dplyr::filter(transcript_id == tx) %>%
    mutate(group = str_to_title(group),
           group = case_when(group == "Star" ~ "STAR",
                             .default = group)) %>%
    ggplot(aes(x = transcript_id, y = log2(counts), fill = exp_group)) +
    #geom_boxplot(colour = "black", outlier.shape = NA) +
    geom_quasirandom(aes(x = exp_group, y = log2(counts),
                         colour = exp_group), size = 3) +
    scale_fill_manual(values = c("royalblue", "darkorange")) +
    scale_colour_manual(values = c("royalblue", "darkorange")) +
    scale_y_continuous(limits = c(-17, 15)) +
    facet_grid(~ group) +
    labs(x = "",
         y = "Transcript Expression",
         fill = "Experimental\nGroup") +
    ggtitle(tx_name) +
    theme_bw() +
    theme(axis.title = element_text(colour = "black", size = 14),
          axis.text.y = element_text(colour = "black", size = 12),
          axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          strip.text = element_text(size = 14, colour = "black"),
          strip.background = element_rect(fill = "lightblue"),
          legend.position = "none")
  # legend.title = element_text(colour = "black", size = 14,
  #                             face = "bold"),
  # legend.text = element_text(colour = "black", size = 12))
  
  
  pcbp2_multimapping <- multimap_df %>%
    dplyr::mutate(group = str_to_title(group),
                  group = case_when(group == "Star" ~ "STAR",
                                    .default = group)) %>%
    dplyr::filter(transcript_id == tx) %>%
    ggplot(aes(x = transcript_id, y = percent_est_from_multimap,
               fill = exp_group)) +
    #geom_boxplot(colour = "black", outlier.shape = NA) +
    geom_quasirandom(aes(x = exp_group, y = percent_est_from_multimap,
                         colour = exp_group), size = 3) +
    scale_y_continuous(limits = c(0, 1)) +
    scale_fill_manual(values = c("royalblue", "darkorange")) +
    scale_colour_manual(values = c("royalblue", "darkorange")) +
    facet_grid(~ group) +
    #ggtitle(tx_name) +
    labs(x = "",
         y = "Multimapped Counts (% of total)",
         fill = "Experimental\nGroup") +
    theme_bw() +
    theme(axis.title = element_text(colour = "black", size = 14),
          axis.text.y = element_text(colour = "black", size = 12),
          axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          strip.text = element_text(size = 14, colour = "black"),
          strip.background = element_rect(fill = "lightblue"),
          legend.title = element_text(colour = "black", size = 14,
                                      face = "bold"),
          legend.text = element_text(colour = "black", size = 12))
  
  pcbp2_legend <- cowplot::get_legend(pcbp2_multimapping)
  
  pcbp2_cowplot <- cowplot::plot_grid(pcbp2_plot, 
                                      pcbp2_multimapping + 
                                        theme(legend.position = "none"),
                                      ncol = 1)
  
  pcbp2_cowplot_legend <- cowplot::plot_grid(pcbp2_cowplot,
                                             pcbp2_legend,
                                             nrow = 1,
                                             rel_widths = c(1, 0.13))
  
  ggsave(plot = pcbp2_cowplot_legend,
         filename = paste0(tx_name, "_cowplot.png"),
         path = here("figures_withSTAR/pcbp2_plots/cowplots/"),
         device = "png",
         height = 225,
         width = 275,
         units = "mm",
         dpi = 400,
         create.dir = TRUE)
}
```

# Classes of Multi-Mapping Counts per Method
```{r}
salmon_mm <- multimap_df %>%
  dplyr::filter(group == "Salmon") %>% 
  dplyr::select(transcript_id,
                ID,
                "salmon_ambig" = AmbigCount,
                "salmon_multimap" = multimap,
                "salmon_percent_multi" = percent_est_from_multimap)

star_mm <- multimap_df %>%
  dplyr::filter(group == "STAR") %>% 
  dplyr::select(transcript_id,
                ID,
                "star_ambig" = AmbigCount,
                "star_multimap" = multimap,
                "star_percent_multi" = percent_est_from_multimap)

sal_star_mm <- left_join(salmon_mm, star_mm, by = c("transcript_id", "ID"))

sal_star_ambig <- sal_star_mm %>%
  ggplot(aes(x = salmon_ambig, y = star_ambig)) +
  geom_point() +
  geom_smooth() +
  facet_wrap(~ ID, ncol = 3) +
  labs(x = "Salmon Ambiguous Counts",
       y = "STAR Ambiguous Counts") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black"),
        strip.background = element_rect(fill = "lightblue"))

ggsave(plot = sal_star_ambig,
       filename = "salmon_star_ambig.png",
       path = here("figures_withSTAR/multi_map/salmon_star_comparison/"),
       device = "png",
       height = 200,
       width = 200,
       units = "mm",
       dpi = 400,
       create.dir = TRUE)

sal_star_ambig
```

```{r}
sal_star_multimap <- sal_star_mm %>%
  ggplot(aes(x = salmon_multimap, y = star_multimap)) +
  geom_point() +
  geom_smooth() +
  facet_wrap(~ ID, ncol = 3) +
  labs(x = "Salmon Multimapping Counts",
       y = "STAR Multimapping Counts") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black"),
        strip.background = element_rect(fill = "lightblue"))

ggsave(plot = sal_star_multimap,
       filename = "salmon_star_multimap.png",
       path = here("figures_withSTAR/multi_map/salmon_star_comparison/"),
       device = "png",
       height = 200,
       width = 200,
       units = "mm",
       dpi = 400,
       create.dir = TRUE)

sal_star_multimap
```

```{r}
sal_star_percent <- sal_star_mm %>%
  ggplot(aes(x = salmon_percent_multi, y = star_percent_multi)) +
  geom_point() +
  geom_smooth() +
  facet_wrap(~ ID, ncol = 3) +
  labs(x = "Salmon Multimapping Counts (% of total counts)",
       y = "SA Multimapping Counts (% of total counts)") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black"),
        strip.background = element_rect(fill = "lightblue"))

ggsave(plot = sal_star_percent,
       filename = "salmon_star_percent.png",
       path = here("figures_withSTAR/multi_map/salmon_star_comparison/"),
       device = "png",
       height = 200,
       width = 200,
       units = "mm",
       dpi = 400,
       create.dir = TRUE)

sal_star_percent
```

# DE Results
## Run DE
```{r}
hisat2_lmfit <- hisat2_dtelist %>%
  de_lmfit()

kallisto_lmfit <- kallisto_dtelist %>% 
  de_lmfit()

salmon_lmfit <- salmon_dtelist %>% 
  de_lmfit()

star_lmfit <- star_dtelist %>% 
  de_lmfit()
```

## Heatmap Plotting of Results
```{r}
pcbp2_ids <- transcript_key %>%
  dplyr::filter(str_detect(transcript_name, "PCBP2-2"))

pcbp2_counts <- full_join(cpm(hisat2_dtelist, log = TRUE) %>%
            set_colnames(paste0("HISAT2_", colnames(hisat2_dtelist))) %>%
            as.data.frame() %>%
            rownames_to_column("transcript_id") %>%
            dplyr::filter(transcript_id %in% pcbp2_ids$transcript_id),
          cpm(kallisto_dtelist, log = TRUE) %>%
            set_colnames(paste0("Kallisto_", colnames(kallisto_dtelist))) %>%
            as.data.frame() %>%
            rownames_to_column("transcript_id") %>%
            dplyr::filter(transcript_id %in% pcbp2_ids$transcript_id),
          by = "transcript_id") %>%
  full_join(cpm(salmon_dtelist, log = TRUE) %>%
              set_colnames(paste0("Salmon_", colnames(salmon_dtelist))) %>%
              as.data.frame() %>%
              rownames_to_column("transcript_id") %>%
              dplyr::filter(transcript_id %in% pcbp2_ids$transcript_id),
            by = "transcript_id") %>%
  full_join(cpm(star_dtelist, log = TRUE) %>%
              set_colnames(paste0("STAR_", colnames(star_dtelist))) %>%
              as.data.frame() %>%
              rownames_to_column("transcript_id") %>%
              dplyr::filter(transcript_id %in% pcbp2_ids$transcript_id),
            by = "transcript_id")
pcbp2_counts[is.na(pcbp2_counts)] <- 0

column_order <- c("HISAT2_SRR13401120", "HISAT2_SRR13401121",
                  "HISAT2_SRR13401122", "HISAT2_SRR13401123",
                  "Kallisto_SRR13401120", "Kallisto_SRR13401121",
                  "Kallisto_SRR13401122", "Kallisto_SRR13401123",
                  "Salmon_SRR13401120", "Salmon_SRR13401121",
                  "Salmon_SRR13401122", "Salmon_SRR13401123",
                  "STAR_SRR13401120", "STAR_SRR13401121",
                  "STAR_SRR13401122", "STAR_SRR13401123",
                  "HISAT2_SRR13401116", "HISAT2_SRR13401117",
                  "HISAT2_SRR13401118", "HISAT2_SRR13401119",
                  "Kallisto_SRR13401116", "Kallisto_SRR13401117",
                  "Kallisto_SRR13401118", "Kallisto_SRR13401119",
                  "Salmon_SRR13401116", "Salmon_SRR13401117",
                  "Salmon_SRR13401118", "Salmon_SRR13401119",
                  "STAR_SRR13401116", "STAR_SRR13401117",
                  "STAR_SRR13401118", "STAR_SRR13401119")

pcbp2_mat <- pcbp2_counts %>%
  left_join(transcript_key,
            by = "transcript_id") %>%
  dplyr::select(-transcript_id) %>%
  column_to_rownames("transcript_name") %>%
  as.matrix()

col_anno <- rep(c("Control", "Knockout"), c(16, 16)) %>%
  as.data.frame() %>%
  set_rownames(column_order) %>%
  set_colnames("Group") %>%
  mutate(Method = rep(
    c(
      rep("HISAT2", 4),
      rep("Kallisto", 4), 
      rep("Salmon", 4),
      rep("STAR", 4)
      ), 2))

hisat2_anno <- hisat2_lmfit %>%
  dplyr::filter(str_detect(transcript_name, "PCBP2-2")) %>%
  dplyr::select(transcript_name, DE) %>%
  as.data.frame() %>%
  mutate(HISAT2 = case_when(
    DE == TRUE ~ "Significant",
    DE == FALSE ~ "Non-Significant"
  )) %>%
  dplyr::select(transcript_name, HISAT2)

kallisto_anno <- kallisto_lmfit %>%
  dplyr::filter(str_detect(transcript_name, "PCBP2-2")) %>%
  dplyr::select(transcript_name, DE) %>%
  as.data.frame() %>%
  mutate(Kallisto = case_when(
    DE == TRUE ~ "Significant",
    DE == FALSE ~ "Non-Significant"
  )) %>%
  dplyr::select(transcript_name, Kallisto)

row_anno <- full_join(hisat2_anno, kallisto_anno, by = "transcript_name")

salmon_anno <- salmon_lmfit %>%
  dplyr::filter(str_detect(transcript_name, "PCBP2-2")) %>%
  dplyr::select(transcript_name, DE) %>%
  as.data.frame() %>%
  mutate(Salmon = case_when(
    DE == TRUE ~ "Significant",
    DE == FALSE ~ "Non-Significant"
  )) %>%
  dplyr::select(transcript_name, Salmon)

row_anno <- full_join(row_anno, salmon_anno, by = "transcript_name")

star_anno <- star_lmfit %>%
  dplyr::filter(str_detect(transcript_name, "PCBP2-2")) %>%
  dplyr::select(transcript_name, DE) %>%
  as.data.frame() %>%
  mutate(STAR = case_when(
    DE == TRUE ~ "Significant",
    DE == FALSE ~ "Non-Significant"
  )) %>%
  dplyr::select(transcript_name, STAR)

row_anno <- full_join(row_anno, star_anno, by = "transcript_name")

row_anno[is.na(row_anno)] <- "Non-Significant"

row_anno <- row_anno %>%
  column_to_rownames("transcript_name")

ann_colours <- list(Group = c(Control = "darkblue",
                              Knockout = "orange"),
                    Method = c(HISAT2 = "black",
                               Kallisto = "brown",
                               Salmon = "salmon",
                               STAR = "pink"),
                    STAR = c(`Non-Significant` = "red",
                           Significant = "green"),
                    Salmon = c(`Non-Significant` = "red",
                               Significant = "green"),
                    HISAT2 = c(`Non-Significant` = "red",
                                Significant = "green"),
                    Kallisto = c(`Non-Significant` = "red",
                                 Significant = "green"))

pcbp2_mat[, column_order] %>%
  pheatmap(border_color = NA,
           cluster_rows = TRUE,
           cluster_cols = FALSE,
           gaps_col = 16,
           annotation_col = col_anno,
           annotation_colors = ann_colours,
           show_colnames = FALSE)

# Use grid.ls(grid.force()) to break the heatmap up into grid elements
# Then steadily turn each element white as you go along until you get the desired layout, using grid.gedit("GRID.text.13423", gp = gpar(col = "white"))
```

# Plot log2 CPM counts
```{r}
pcbp2_df <- pcbp2_mat %>%
  as.data.frame() %>%
  rownames_to_column("transcript_name") %>%
  pivot_longer(cols = !starts_with("t"),
               names_to = "ID",
               values_to = "logCPM") %>%
  dplyr::mutate("sample_id" = str_remove(ID, ".*_"),
                "Method" = str_remove(ID, "_.*")) %>%
  left_join(dplyr::select(hisat2_dtelist$samples,
                          "sample_id" = "ID", "Group" = "group"), 
            by = "sample_id") %>%
  dplyr::mutate(Group = str_to_title(Group),
                sample_id = factor(sample_id, 
                                   levels = c("SRR13401120", "SRR13401121",
                                              "SRR13401122", "SRR13401123",
                                              "SRR13401116", "SRR13401117",
                                              "SRR13401118",
                                              "SRR13401119")))

for(i in unique(pcbp2_df$transcript_name)) {
  pcbp2_plot <- pcbp2_df %>%
    dplyr::filter(transcript_name == i) %>%
    ggplot(aes(x = sample_id, y = logCPM, colour = Group)) +
    geom_point(size = 3) +
    labs(x = "",
         y = "Transcript Expression (log2 CPM)") +
    scale_colour_manual(values = c("darkblue", "orange")) +
    ggtitle(i) +
    facet_wrap(~ Method, nrow = 1, ncol = 4) +
    labs(x = "",
         y = "Transcript Expression (log2 CPM)") +
    theme_bw() +
    theme(axis.text.x = element_blank(),
          axis.text.y = element_text(colour = "black", size = 12),
          axis.ticks.x = element_blank(),
          axis.title.y = element_text(colour = "black", size = 14),
          strip.background = element_rect(fill = "lightblue"),
          strip.text = element_text(colour = "black", size = 14),
          legend.title = element_text(colour = "black", size = 14),
          legend.text = element_text(colour = "black", size = 12),
          plot.title = element_text(colour = "black", size = 14,
                                    face = "bold"))
  
  ggsave(plot = pcbp2_plot,
         filename = paste0(i, "_log2_cpm.png"),
         path = here("figures_withSTAR/pcbp2_plots/logcpm_counts/"),
         device = "png",
         height = 150,
         width = 200,
         units = "mm",
         dpi = 400,
         create.dir = TRUE)
}

pcbp2_plot
```

## log2 CPM cowplots with Multi-Mapping
```{r}

for(tx in unique(pcbp2_df$transcript_name)) {
  pcbp2_plot <- pcbp2_df %>%
    dplyr::filter(transcript_name == tx) %>%
    ggplot(aes(x = sample_id, y = logCPM, colour = Group)) +
    geom_point(size = 3) +
    labs(x = "",
         y = "Transcript Expression (log2 CPM)") +
    scale_colour_manual(values = c("darkblue", "orange")) +
    ggtitle(tx) +
    facet_wrap(~ Method, nrow = 1, ncol = 4) +
    labs(x = "",
         y = "Transcript Expression (log2 CPM)") +
    theme_bw() +
    theme(axis.text.x = element_blank(),
          axis.text.y = element_text(colour = "black", size = 12),
          axis.ticks.x = element_blank(),
          axis.title.y = element_text(colour = "black", size = 14),
          strip.background = element_rect(fill = "lightblue"),
          strip.text = element_text(colour = "black", size = 14),
          legend.title = element_text(colour = "black", size = 14),
          legend.text = element_text(colour = "black", size = 12),
          plot.title = element_text(colour = "black", size = 14,
                                    face = "bold"),
          legend.position = "none")

  pcbp2_multimapping <- multimap_df %>%
    dplyr::mutate(group = str_to_title(group),
                  group = case_when(group == "Star" ~ "STAR",
                                    .default = group)) %>%
    dplyr::filter(transcript_name == tx) %>%
    dplyr::select(transcript_name, percent_est_from_multimap,
                  exp_group, group) %>%
    rbind(c("PCBP2-203", -0.05, "Control", "Kallisto")) %>%
    mutate(
      percent_est_from_multimap = as.numeric(percent_est_from_multimap)
      ) %>%
    ggplot(aes(x = exp_group, y = 100*percent_est_from_multimap,
               colour = exp_group)) +
    geom_quasirandom(size = 3) +
    scale_y_continuous(limits = c(0, 100)) +
    scale_colour_manual(values = c("darkblue", "darkorange")) +
    facet_grid(~ group) +
    labs(x = "",
         y = "Multimapped Counts (% of total)",
         colour = "Experimental\nGroup") +
    theme_bw() +
    theme(axis.title = element_text(colour = "black", size = 14),
          axis.text.y = element_text(colour = "black", size = 12),
          axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          strip.text = element_text(size = 14, colour = "black"),
          strip.background = element_rect(fill = "lightblue"),
          legend.title = element_text(colour = "black", size = 14,
                                      face = "bold"),
          legend.text = element_text(colour = "black", size = 12),
          legend.position = "right")
  
  pcbp2_legend <- cowplot::get_legend(pcbp2_multimapping)
  
  blanc <- ggplot() +
    geom_blank() +
    theme_void()
  
  pcbp2_cowplot_a <- cowplot::plot_grid(blanc,
                                        pcbp2_plot, 
                                        ncol = 2,
                                        rel_widths = c(0.025, 1))
  
  pcbp2_cowplot_adjust <- cowplot::plot_grid(pcbp2_cowplot_a, 
                                      pcbp2_multimapping + 
                                        theme(legend.position = "none"),
                                      nrow = 2)
  
  pcbp2_cowplot_legend <- cowplot::plot_grid(pcbp2_cowplot_adjust,
                                             pcbp2_legend,
                                             nrow = 1,
                                             rel_widths = c(1, 0.13))
  
  ggsave(plot = pcbp2_cowplot_legend,
         filename = paste0(tx, "_logcpm_cowplot.png"),
         path = here("figures_withSTAR/pcbp2_plots/cowplots_logcpm/"),
         device = "png",
         height = 225,
         width = 275,
         units = "mm",
         dpi = 400,
         create.dir = TRUE)
}

pcbp2_cowplot_legend
```
