---
title: "DTE and DTU kmer analysis"
author: "Justin Bogias"
date: "2024-08-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(engine.opts = list(bash = "-l"))
```

# Load packages
First we will load in the packages needed to execute this analysis
```{r}
library(dplyr)
library(readr)
library(tidyr)
library(stringr)
library(magrittr)
library(ggplot2)
library(viridis)
library(ggfortify)
library(PCAtools)
library(edgeR)
library(AnnotationHub)
library(ensembldb)
library(pheatmap)
library(cowplot)
library(ggbeeswarm)
library(tibble)
library(EnsDb.Hsapiens.v86)
library(GenomicFeatures)
library(here)
```

## Load Custom Functions
```{r}
source(here("R/cor_funcs_variance.R"))
```

Some methods will be called "hisat2", this just means that bootstrapping estimates have not been taken into account. Nothing super special going on.

I have also defined several functions here for quality of life purposes
```{r}
# From 01_Annotations
gene_txp_anno <- read_csv(here("data/annotations/grch38_103_df.csv.gz"))

transcript_key <- gene_txp_anno %>%
  dplyr::filter(type == "transcript") %>%
  dplyr::select(transcript_id,
                transcript_name)

txp_gene_ensdb_lengths <- read_csv(
  here("data/annotations/txp_gene_ensdb_lengths.csv.gz")
  )

transcript_key <- dplyr::select(txp_gene_ensdb_lengths,
                                c("transcript_id" = tx_id,
                                  "transcript_name" = tx_name))

# From 02_dtelist_prep
hisat2_dtelist <- read_rds(here("data/dtelists/hisat2_dtelist.rds"))
kallisto_dtelist <- read_rds(here("data/dtelists/kallisto_dtelist.rds"))
sa_dtelist <- read_rds(here("data/dtelists/sa_dtelist.rds"))
star_dtelist <- read_rds(here("data/dtelists/star_dtelist.rds"))

# From 04_principal_component_analysis
hisat2_loadings <- read_csv(
  here("data/pca/loadings/hisat2_loadings.csv")
  ) %>%
  head(250)

kallisto_loadings <- read_csv(
  here("data/pca/loadings/kallisto_loadings.csv")
  ) %>%
  head(275)

salmon_loadings <- read_csv(
  here("data/pca/loadings/salmon_loadings.csv")
  ) %>%
  head(300)

star_loadings <- read_csv(
  here("data/pca/loadings/star_loadings.csv")
  ) %>%
  head(275)

# From 05_DTE_DTU
# From 05_dte_and_dtu
## DTE
hisat2_dte_exclusive <- read_csv(
  here("data/differential_expression/hisat2_exclusive_tx.csv")
  )

kallisto_dte_exclusive <- read_csv(
  here("data/differential_expression/kallisto_exclusive_tx.csv")
  ) %>%
  head(28)

salmon_dte_exclusive <- read_csv(
  here("data/differential_expression/salmon_exclusive_tx.csv")
  ) %>%
  head(28)

star_dte_exclusive <- read_csv(
  here("data/differential_expression/star_exclusive_tx.csv")
  ) %>%
  head(28)

all_methods_tx_dte <- read_csv(
  here("data/differential_expression/all_methods_tx.csv")
) %>%
  head(19)

## DTU
hisat2_dtu_exclusive <- read_csv(
  here("data/differential_usage/hisat2_exclusive_dtu.csv")
  ) %>%
  dplyr::rename("transcript_id" = Transcript) %>%
  dplyr::arrange(txpFDR) %>%
  head(100)

kallisto_dtu_exclusive <- read_csv(
  here("data/differential_usage/kallisto_exclusive_dtu.csv")
  ) %>%
  dplyr::rename("transcript_id" = Transcript) %>%
  dplyr::arrange(txpFDR) %>%
  head(100)

star_dtu_exclusive <- read_csv(
  here("data/differential_usage/star_exclusive_dtu.csv")
  ) %>%
  dplyr::rename("transcript_id" = Transcript) %>%
  dplyr::arrange(txpFDR) %>%
  head(100)

salmon_dtu_exclusive <- read_csv(
  here("data/differential_usage/salmon_exclusive_dtu.csv")
  ) %>%
  dplyr::rename("transcript_id" = Transcript) %>%
  dplyr::arrange(txpFDR) %>%
  head(100)

all_methods_tx_dtu <- read_csv(
  here("data/differential_usage/all_methods_tx.csv")
  ) %>%
  dplyr::rename("transcript_id" = Transcript) %>%
  dplyr::arrange(txpFDR) %>%
  head(100)
```

# Get Combined Data
```{r cor_df}
joined_df <- get_joined_samples(w = hisat2_dtelist,
                                x = kallisto_dtelist,
                                y = star_dtelist,
                                z = sa_dtelist,
                                method_w = "HISAT2",
                                method_x = "Kallisto",
                                method_y = "STAR",
                                method_z = "Salmon",
                                sample = 1)

cor_df <- joined_df
cor_df$variance <- apply(cor_df[,-1], 1, var)
```


```{r whole_cor_df}
whole_cor <- cor_df %>%
  dplyr::left_join(gene_txp_anno,
                   by = "transcript_id")
```

# Kmer Analysis by DTE analysis
## hisat2
```{r hisat2_dte_kmer}
bt2_tx_remove <- txp_gene_ensdb_lengths %>%
  dplyr::filter(tx_id %in% hisat2_dte_exclusive$transcript_id) %>%
  dplyr::filter(str_detect(seqnames, "^CHR")) %>%
  as.data.frame()

hisat2_dte_exclusive <- hisat2_dte_exclusive %>% 
  dplyr::filter(!transcript_id %in% bt2_tx_remove$tx_id)

yTx_hisat2_dte <- exonsBy(
  edb,
  filter = TxIdFilter(hisat2_dte_exclusive$transcript_id)
  )

yTxSeqs_hisat2_dte <- extractTranscriptSeqs(dna, yTx_hisat2_dte)

hisat2_dte_kmer_df <- yTxSeqs_hisat2_dte %>%
  as.data.frame() %>%
  rownames_to_column("transcript_id") %>%
  mutate(transcript_id = paste0(">", transcript_id)) %>%
  set_colnames(c("transcript_id", "sequence")) %>%
  as_tibble()

hisat2_dte_kmer_df %>%
  head(200) %>%
  write_delim(here("data/kmer_analysis/loadings/hisat2_dte/hisat2_seqs.txt"),
              delim = "\n",
              eol = "\n\n",
              col_names = FALSE)
```

```{sh eval=FALSE}
loc=/Users/konstantinosbogias/Documents/PhD/R_Projects/MethodsChapter/data/kmer_analysis

cp ${loc}/hisat2_dte/hisat2_seqs.txt ${loc}/hisat2_dte/hisat2_seqs.fa
for i in {1..100}; do
    jellyfish count -m ${i} -s 100M --output ${loc}/hisat2_dte/hisat2_mer_counts.jf -C ${loc}/hisat2_dte/hisat2_seqs.fa
    jellyfish histo ${loc}/hisat2_dte/hisat2_mer_counts.jf > ${loc}/hisat2_dte/histos/hisat2_${loc}_mer_counts_histo.txt
    jellyfish stats ${loc}/hisat2_dte/hisat2_mer_counts.jf > ${loc}/hisat2_dte/stats/hisat2_${i}_mer_counts_stats.txt
    echo "hisat2 ${i}_mer done"
done
```

### Kmer data
```{r kmer_plots_hisat2 eval=TRUE}
hisat2_dte_mer_hist <- data.frame()
for(file in list.files(here("data/kmer_analysis/loadings/hisat2_dte/histos"))) {
  hisat2_dte_mer_add <- read_delim(
    paste0(here("data/kmer_analysis/loadings/hisat2_dte/histos/"), file),
    col_names = FALSE,
    delim = " ") %>%
    set_colnames(c("Repeats", "Frequency")) %>%
    mutate(kmer_length = file %>%
             str_extract("...mer") %>%
             str_remove(".*_") %>% 
             str_remove("mer") %>% 
             as.numeric())
  
  hisat2_dte_mer_hist <- rbind(hisat2_dte_mer_hist, hisat2_dte_mer_add)
}

hisat2_dte_mer_hist <- hisat2_dte_mer_hist %>%
  mutate(group = "hisat2")

write_csv(hisat2_dte_mer_hist,
          here("data/kmer_analysis/loadings/hisat2_dte/hisat2_mer_hist.csv"))

hisat2_dte_mer_stat <- data.frame()
for(file in list.files(here("data/kmer_analysis/loadings/hisat2_dte/stats"))) {
  hisat2_dte_stat_add <- read_delim(
    paste0(here("data/kmer_analysis/loadings/hisat2_dte/stats/"), file),
    col_names = FALSE,
    delim = " ") %>% 
    as.data.frame() %>%
    column_to_rownames("X1") %>%
    mutate_if(is.character, as.numeric) %>%
    mutate(Frequency = coalesce(X2, X3, X4, X5)) %>%
    rownames_to_column("Type") %>%
    dplyr::select("Type", "Frequency") %>%
    mutate(Type = str_remove(Type, ":")) %>% 
    dplyr::mutate(kmer_length = file %>% 
                    str_extract("...mer") %>% 
                    str_remove(".*_") %>% 
                    str_remove("mer") %>%
                    as.numeric())
  
  hisat2_dte_mer_stat <- rbind(hisat2_dte_mer_stat, hisat2_dte_stat_add)
}

hisat2_dte_mer_stat <- hisat2_dte_mer_stat %>%
  mutate(group = "hisat2")

write_csv(hisat2_dte_mer_stat,
          here("data/kmer_analysis/loadings/hisat2_dte/hisat2_mer_stat.csv"))
```

### Kmer plotting
Wanna try and find long genes with less distinct kmers, so low Distinct and Unique, with a high Total. Possibly a low Max count but a high Repeat number.
```{r}
hisat2_dte_lengths <- txp_gene_ensdb_lengths %>%
  dplyr::filter(tx_id %in% hisat2_dte_exclusive$transcript_id) %>%
  dplyr::select(tx_id, transcript_length, seqnames, strand, tx_biotype) %>%
  dplyr::mutate(group = "hisat2")
hisat2_dte_lengths$transcript_length %>% summary()

# Summarise kmer lengths
hisat2_dte_mer_hist$kmer_length %>% table() %>% as.data.frame()

hisat2_dte_mer_hist %>%
  dplyr::filter(kmer_length >= 31) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ as.character(Repeats),
             scales = "free") +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"))

hisat2_dte_mer_hist %>%
  dplyr::filter(kmer_length <= 20) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"),
        legend.position = "none")

hisat2_dte_mer_stat %>%
  dplyr::filter(!Type == "Max_count") %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = Type)) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ Type) +
  theme_bw()
```

## Kallisto
```{r kallisto_kmer}
k_tx_remove <- txp_gene_ensdb_lengths %>%
  dplyr::filter(tx_id %in% kallisto_dte_exclusive$transcript_id) %>%
  dplyr::filter(str_detect(seqnames, "^CHR")) %>%
  as.data.frame()

kallisto_dte_exclusive <- kallisto_dte_exclusive %>%
  dplyr::filter(transcript_id %in% gene_txp_anno$transcript_id)

yTx_kallisto_dte <- exonsBy(
  edb, 
  filter = TxIdFilter(kallisto_dte_exclusive$transcript_id)
  )

yTxSeqs_kallisto_dte <- extractTranscriptSeqs(dna, yTx_kallisto_dte)

kallisto_dte_kmer_df <- yTxSeqs_kallisto_dte %>%
  as.data.frame() %>%
  rownames_to_column("transcript_id") %>%
  mutate(transcript_id = paste0(">", transcript_id)) %>%
  set_colnames(c("transcript_id", "sequence")) %>%
  as_tibble()

kallisto_dte_kmer_df %>%
  write_delim(here("data/kmer_analysis/loadings/kallisto_dte/kallisto_seqs.txt"),
              delim = "\n",
              eol = "\n\n",
              col_names = FALSE)
```

```{sh eval=FALSE}
loc=/Users/konstantinosbogias/Documents/PhD/R_Projects/MethodsChapter/data/kmer_analysis

cp ${loc}/kallisto_dte/kallisto_seqs.txt ${loc}/kallisto_dte/kallisto_seqs.fa
for i in {1..100}; do
    jellyfish count -m ${i} -s 100M --output ${loc}/kallisto_dte/kallisto_mer_counts.jf -C ${loc}/kallisto_dte/kallisto_seqs.fa
    jellyfish histo ${loc}/kallisto_dte/kallisto_mer_counts.jf > ${loc}/kallisto_dte/histos/kallisto_${i}_mer_counts_histo.txt
    jellyfish stats ${loc}/kallisto_dte/kallisto_mer_counts.jf > ${loc}/kallisto_dte/stats/kallisto_${i}_mer_counts_stats.txt
    echo "kallisto ${i}_mer done"
done
```

### Kmer data
```{r kmer_plots_kallisto eval=TRUE}
kallisto_dte_mer_hist <- data.frame()
for(file in list.files(here("data/kmer_analysis/loadings/kallisto_dte/histos"))) {
  kallisto_dte_mer_add <- read_delim(
    paste0(here("data/kmer_analysis/loadings/kallisto_dte/histos/"), file),
    col_names = FALSE,
    delim = " ") %>%
    set_colnames(c("Repeats", "Frequency")) %>%
    mutate(kmer_length = file %>%
             str_extract("...mer") %>%
             str_remove(".*_") %>% 
             str_remove("mer") %>% 
             as.numeric())
  
  kallisto_dte_mer_hist <- rbind(kallisto_dte_mer_hist, kallisto_dte_mer_add)
}

kallisto_dte_mer_hist <- kallisto_dte_mer_hist %>%
  mutate(group = "kallisto")

write_csv(kallisto_dte_mer_hist,
          here("data/kmer_analysis/loadings/kallisto_dte/kallisto_mer_hist.csv"))

kallisto_dte_mer_stat <- data.frame()
for(file in list.files(here("data/kmer_analysis/loadings/kallisto_dte/stats"))) {
  kallisto_dte_stat_add <- read_delim(
    paste0(here("data/kmer_analysis/loadings/kallisto_dte/stats/"), file),
    col_names = FALSE,
    delim = " ") %>% 
    as.data.frame() %>%
    column_to_rownames("X1") %>%
    mutate_if(is.character, as.numeric) %>%
    mutate(Frequency = coalesce(X2, X3, X4, X5)) %>%
    rownames_to_column("Type") %>%
    dplyr::select("Type", "Frequency") %>%
    mutate(Type = str_remove(Type, ":")) %>% 
    dplyr::mutate(kmer_length = file %>% 
                    str_extract("...mer") %>% 
                    str_remove(".*_") %>% 
                    str_remove("mer") %>%
                    as.numeric())
  
  kallisto_dte_mer_stat <- rbind(kallisto_dte_mer_stat, kallisto_dte_stat_add)
}

kallisto_dte_mer_stat <- kallisto_dte_mer_stat %>%
  mutate(group = "kallisto")

write_csv(kallisto_dte_mer_stat,
          here("data/kmer_analysis/loadings/kallisto_dte/kallisto_mer_stat.csv"))
```

### Kmer plotting
Wanna try and find long genes with less distinct kmers, so low Distinct and Unique, with a high Total. Possibly a low Max count but a high Repeat number.
```{r}
kallisto_dte_lengths <- txp_gene_ensdb_lengths %>%
  dplyr::filter(tx_id %in% kallisto_dte_exclusive$transcript_id) %>%
  dplyr::select(tx_id, transcript_length, seqnames, strand, tx_biotype) %>%
  dplyr::mutate(group = "kallisto")
kallisto_dte_lengths$transcript_length %>% summary()

# Summarise kmer lengths
kallisto_dte_mer_hist$kmer_length %>% table() %>% as.data.frame()

kallisto_dte_mer_hist %>%
  dplyr::filter(kmer_length >= 31) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ as.character(Repeats),
             scales = "free") +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"))

kallisto_dte_mer_hist %>%
  dplyr::filter(kmer_length <= 20) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"),
        legend.position = "none")

kallisto_dte_mer_stat %>%
  dplyr::filter(!Type == "Max_count") %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = Type)) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ Type) +
  theme_bw()
```

## Salmon
```{r star_kmer}
salmon_dte_tx_remove <- txp_gene_ensdb_lengths %>%
  dplyr::filter(tx_id %in% salmon_dte_exclusive$transcript_id) %>%
  dplyr::filter(str_detect(seqnames, "^CHR")) %>%
  as.data.frame()

salmon_dte_exclusive <- salmon_dte_exclusive %>%
  dplyr::filter(!transcript_id %in% salmon_dte_tx_remove$tx_id)

yTx_salmon_dte <- exonsBy(edb,
                          filter = TxIdFilter(salmon_dte_exclusive$transcript_id))

yTxSeqs_salmon_dte <- extractTranscriptSeqs(dna, yTx_salmon_dte)

salmon_kmer_df <- yTxSeqs_salmon_dte %>%
  as.data.frame() %>%
  rownames_to_column("transcript_id") %>%
  mutate(transcript_id = paste0(">", transcript_id)) %>%
  set_colnames(c("transcript_id", "sequence")) %>%
  as_tibble()

salmon_kmer_df %>%
  head(19) %>%
  write_delim(here("data/kmer_analysis/loadings/salmon_dte/salmon_seqs.txt"),
              delim = "\n",
              eol = "\n\n",
              col_names = FALSE)
```

## STAR
```{r star_kmer}
star_dte_tx_remove <- txp_gene_ensdb_lengths %>%
  dplyr::filter(tx_id %in% star_dte_exclusive$transcript_id) %>%
  dplyr::filter(str_detect(seqnames, "^CHR")) %>%
  as.data.frame()

star_dte_exclusive <- star_dte_exclusive %>%
  dplyr::filter(!transcript_id %in% star_dte_tx_remove$tx_id)

yTx_star_dte <- exonsBy(
  edb,
  filter = TxIdFilter(star_dte_exclusive$transcript_id)
  )

yTxSeqs_star_dte <- extractTranscriptSeqs(dna, yTx_star_dte)

star_kmer_df <- yTxSeqs_star_dte %>%
  as.data.frame() %>%
  rownames_to_column("transcript_id") %>%
  mutate(transcript_id = paste0(">", transcript_id)) %>%
  set_colnames(c("transcript_id", "sequence")) %>%
  as_tibble()

star_kmer_df %>%
  head(19) %>%
  write_delim(here("data/kmer_analysis/loadings/star_dte/star_seqs.txt"),
              delim = "\n",
              eol = "\n\n",
              col_names = FALSE)
```

```{sh eval=FALSE}
loc=/Users/konstantinosbogias/Documents/PhD/R_Projects/MethodsChapter/data/kmer_analysis

cp ${loc}/star_dte/star_seqs.txt ${loc}/star_dte/star_seqs.fa
for i in {1..100}; do
    jellyfish count -m ${i} -s 100M --output ${loc}/star_dte/star_mer_counts.jf -C ${loc}/star_dte/sal_sal_seqs.fa
    jellyfish histo ${loc}/star_dte/star_mer_counts.jf > ${loc}/star_dte/histos/star_${i}_mer_counts_histo.txt
    jellyfish stats ${loc}/star_dte/star_mer_counts.jf > ${loc}/star_dte/stats/star_${i}_mer_counts_stats.txt
    echo "sa-salmon ${i}_mer done"
done
```

### Kmer data
```{r kmer_plots_salmon eval=TRUE}
star_dte_mer_hist <- data.frame()
for(file in list.files(here("data/kmer_analysis/loadings/star_dte/histos"))) {
  star_dte_mer_add <- read_delim(
    paste0(here("data/kmer_analysis/loadings/star_dte/histos/"), file),
    col_names = FALSE,
    delim = " ") %>%
    set_colnames(c("Repeats", "Frequency")) %>%
    mutate(kmer_length = file %>%
             str_extract("...mer") %>%
             str_remove(".*_") %>% 
             str_remove("mer") %>% 
             as.numeric())
  
  star_dte_mer_hist <- rbind(star_dte_mer_hist, star_dte_mer_add)
}

star_dte_mer_hist <- star_dte_mer_hist %>%
  mutate(group = "sa-salmon")

write_csv(star_dte_mer_hist,
          here("data/kmer_analysis/loadings/star_dte/star_dte_mer_hist.csv"))

star_dte_mer_stat <- data.frame()
for(file in list.files(here("data/kmer_analysis/loadings/star_dte/stats"))) {
  star_dte_stat_add <- read_delim(
    paste0(here("data/kmer_analysis/loadings/star_dte/stats/"), file),
    col_names = FALSE,
    delim = " ") %>% 
    as.data.frame() %>%
    column_to_rownames("X1") %>%
    mutate_if(is.character, as.numeric) %>%
    mutate(Frequency = coalesce(X2, X3, X4, X5)) %>%
    rownames_to_column("Type") %>%
    dplyr::select("Type", "Frequency") %>%
    mutate(Type = str_remove(Type, ":")) %>% 
    dplyr::mutate(kmer_length = file %>% 
                    str_extract("...mer") %>% 
                    str_remove(".*_") %>% 
                    str_remove("mer") %>%
                    as.numeric())
  
  star_dte_mer_stat <- rbind(star_dte_mer_stat, star_dte_stat_add)
}

star_dte_mer_stat <- star_dte_mer_stat %>%
  mutate(group = "sa-sal")

write_csv(star_dte_mer_stat,
          here("data/kmer_analysis/loadings/star_dte/star_dte_mer_stat.csv"))
```

### Kmer plotting
Wanna try and find long genes with less distinct kmers, so low Distinct and Unique, with a high Total. Possibly a low Max count but a high Repeat number.
```{r}
star_dte_lengths <- txp_gene_ensdb_lengths %>%
  dplyr::filter(tx_id %in% star_dte_exclusive$transcript_id) %>%
  dplyr::select(tx_id, transcript_length, seqnames, strand, tx_biotype) %>%
  dplyr::mutate(group = "sa-sal")
star_dte_lengths$transcript_length %>% summary()

# Summarise kmer lengths
star_dte_mer_hist$kmer_length %>% table() %>% as.data.frame()

star_dte_mer_hist %>%
  dplyr::filter(kmer_length >= 31) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ as.character(Repeats),
             scales = "free") +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"))

star_dte_mer_hist %>%
  dplyr::filter(kmer_length <= 20) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"),
        legend.position = "none")

star_dte_mer_stat %>%
  dplyr::filter(!Type == "Max_count") %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = Type)) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ Type) +
  theme_bw()
```

## All Methods DTE
```{r star_kmer}
all_methods_dte_tx_remove <- txp_gene_ensdb_lengths %>%
  dplyr::filter(tx_id %in% all_methods_tx_dte$transcript_id) %>%
  dplyr::filter(str_detect(seqnames, "^CHR")) %>%
  as.data.frame()

all_methods_tx_dte <- all_methods_tx_dte %>%
  dplyr::filter(!transcript_id %in% all_methods_dte_tx_remove$tx_id)

yTx_all_methods_dte <- exonsBy(
  edb,
  filter = TxIdFilter(all_methods_tx_dte$transcript_id)
  )

yTxSeqs_all_methods_dte <- extractTranscriptSeqs(dna, yTx_all_methods_dte)

all_methods_dte_kmer_df <- yTxSeqs_all_methods_dte %>%
  as.data.frame() %>%
  rownames_to_column("transcript_id") %>%
  mutate(transcript_id = paste0(">", transcript_id)) %>%
  set_colnames(c("transcript_id", "sequence")) %>%
  as_tibble()

all_methods_dte_kmer_df %>%
  head(19) %>%
  write_delim(here("data/kmer_analysis/all_methods_dte/all_methods_seqs.txt"),
              delim = "\n",
              eol = "\n\n",
              col_names = FALSE)
```

```{sh eval=FALSE}
loc=/Users/konstantinosbogias/Documents/PhD/R_Projects/MethodsChapter/data/kmer_analysis

cp ${loc}/all_methods_dte/all_methods_seqs.txt ${loc}/all_methods_dte/all_methods_seqs.fa
for i in {1..100}; do
    jellyfish count -m ${i} -s 100M --output ${loc}/all_methods_dte/all_methods_mer_counts.jf -C ${loc}/all_methods_dte/all_methods_seqs.fa
    jellyfish histo ${loc}/all_methods_dte/all_methods_mer_counts.jf > ${loc}/all_methods_dte/histos/all_methods_${i}_mer_counts_histo.txt
    jellyfish stats ${loc}/all_methods_dte/all_methods_mer_counts.jf > ${loc}/all_methods_dte/stats/all_methods_${i}_mer_counts_stats.txt
    echo "all_methods ${i}_mer done"
done
```

### Kmer data
```{r kmer_plots_salmon eval=TRUE}
all_methods_dte_mer_hist <- data.frame()
for(file in list.files(here("data/kmer_analysis/all_methods_dte/histos"))) {
  all_methods_dte_mer_add <- read_delim(
    paste0(here("data/kmer_analysis/all_methods_dte/histos/"), file),
    col_names = FALSE,
    delim = " ") %>%
    set_colnames(c("Repeats", "Frequency")) %>%
    mutate(kmer_length = file %>%
             str_extract("...mer") %>%
             str_remove(".*_") %>% 
             str_remove("mer") %>% 
             as.numeric())
  
  all_methods_dte_mer_hist <- rbind(all_methods_dte_mer_hist,
                                    all_methods_dte_mer_add)
}

all_methods_dte_mer_hist <- all_methods_dte_mer_hist %>%
  mutate(group = "all_methods")

write_csv(all_methods_dte_mer_hist,
          here("data/kmer_analysis/all_methods_dte/all_methods_dte_mer_hist.csv"))

all_methods_dte_mer_stat <- data.frame()
for(file in list.files(here("data/kmer_analysis/all_methods_dte/stats"))) {
  all_methods_dte_stat_add <- read_delim(
    paste0(here("data/kmer_analysis/all_methods_dte/stats/"), file),
    col_names = FALSE,
    delim = " ") %>% 
    as.data.frame() %>%
    column_to_rownames("X1") %>%
    mutate_if(is.character, as.numeric) %>%
    mutate(Frequency = coalesce(X2, X3, X4, X5)) %>%
    rownames_to_column("Type") %>%
    dplyr::select("Type", "Frequency") %>%
    mutate(Type = str_remove(Type, ":")) %>% 
    dplyr::mutate(kmer_length = file %>% 
                    str_extract("...mer") %>% 
                    str_remove(".*_") %>% 
                    str_remove("mer") %>%
                    as.numeric())
  
  all_methods_dte_mer_stat <- rbind(all_methods_dte_mer_stat,
                                    all_methods_dte_stat_add)
}

all_methods_dte_mer_stat <- all_methods_dte_mer_stat %>%
  mutate(group = "all_methods")

write_csv(all_methods_dte_mer_stat,
          here("data/kmer_analysis/all_methods_dte/all_methods_dte_mer_stat.csv"))
```

### Kmer plotting
Wanna try and find long genes with less distinct kmers, so low Distinct and Unique, with a high Total. Possibly a low Max count but a high Repeat number.
```{r}
all_methods_dte_lengths <- txp_gene_ensdb_lengths %>%
  dplyr::filter(tx_id %in% all_methods_tx_dte$transcript_id) %>%
  dplyr::select(tx_id, transcript_length, seqnames, strand, tx_biotype) %>%
  dplyr::mutate(group = "all_methods")
all_methods_dte_lengths$transcript_length %>% summary()

# Summarise kmer lengths
all_methods_dte_mer_hist$kmer_length %>% table() %>% as.data.frame()

all_methods_dte_mer_hist %>%
  dplyr::filter(kmer_length >= 31) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ as.character(Repeats),
             scales = "free") +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"))

all_methods_dte_mer_hist %>%
  dplyr::filter(kmer_length <= 20) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"),
        legend.position = "none")

all_methods_dte_mer_stat %>%
  dplyr::filter(!Type == "Max_count") %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = Type)) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ Type) +
  theme_bw()
```

## DTE Plotting
```{r}
lengths_df <- rbind(hisat2_dte_lengths,
                    kallisto_dte_lengths,
                    star_dte_lengths,
                    all_methods_dte_lengths)

lengths_by_method <- lengths_df %>%
  dplyr::mutate(
    group = case_when(
      group == "hisat2" ~ "HISAT2 Exclusive",
      group == "kallisto" ~ "Kallisto Exclusive",
      group == "salmon" ~ "Salmon Exclusive",
      group == "star" ~ "STAR Exclusive",
      group == "all_methods" ~ "All Methods"
    ),
    group = factor(group, levels = c("All Methods",
                                     "HISAT2 Exclusive",
                                     "Kallisto Exclusive",
                                     "Salmon Exclusive",
                                     "STAR Exclusive"))) %>%
  ggplot(aes(x = group, y = transcript_length)) +
  geom_boxplot(fill = "lightblue") +
  scale_y_log10() +
  labs(x = "",
       y = "Transcript Length (log10)") +
  theme_bw()

lengths_by_method %>%
  ggsave(filename = "length_by_method_plot.png",
         path = here("figures/kmer_analysis/dte_kmer_plots/"),
         device = "png",
         height = 200,
         width = 200,
         units = "mm",
         dpi = 400)

txp_lengths_hist <- txp_gene_ensdb_lengths %>%
  dplyr::select(tx_id, transcript_length) %>%
  mutate(group = "All Transcripts") %>%
  rbind(lengths_df %>%
          dplyr::select(tx_id, transcript_length) %>%
          dplyr::mutate(group = "Top Loadings Transcripts")) %>%
  ggplot(aes(x = log10(transcript_length))) +
  geom_histogram(bins = 200) +
  scale_x_continuous(limits = c(1, 6),
                     breaks = c(1, 2, 3, 4, 5, 6)) +
  labs(x = "Transcript Length (log10)",
       y = "Frequency") +
  facet_grid(group ~ ., switch = "y", scales = "free") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 12),
        axis.title = element_text(colour = "black", size = 14),
        strip.background = element_rect(fill = "lightblue"),
        strip.text = element_text(colour = "black", size = 14))

txp_lengths_hist %>%
  ggsave(filename = "length_plot.png",
         path = here("figures/kmer_analysis/dte_kmer_plots/"),
         device = "png",
         height = 150,
         width = 200,
         units = "mm",
         dpi = 400)

# We can see that there is stuff all variation in frequency within groups
stat_df <- rbind(hisat2_dte_mer_stat,
                 kallisto_dte_mer_stat,
                 star_dte_mer_stat,
                 all_methods_dte_mer_stat)

group_names <- stat_df$group %>% unique()
new_stat_df <- data.frame()
stat_group_chunk <- data.frame()
for(g in group_names) {
  for(i in 1:100) {
    stat_kmer_chunk <- stat_df %>%
      dplyr::filter(kmer_length == i) %>%
      dplyr::filter(group == g) %>%
      dplyr::select(Type, Frequency) %>%
      column_to_rownames("Type") %>%
      t() %>%
      as_tibble() %>%
      mutate(Multiplicity = Total-Distinct) %>%
      t() %>%
      set_colnames("Frequency") %>%
      as.data.frame() %>%
      rownames_to_column("Type") %>%
      mutate(kmer_length = i, group = g)
    
    stat_group_chunk <- rbind(stat_group_chunk, stat_kmer_chunk)
  }
  new_stat_df <- rbind(new_stat_df, stat_group_chunk)
}


kmer_stat_plotting <- function(x, type, kmer = 31) {
  x %>%
    dplyr::filter(kmer_length >= kmer,
                  Type == type) %>%
    dplyr::mutate(
      group = case_when(
        group == "hisat2" ~ "HISAT2 Exclusive",
        group == "kallisto" ~ "Kallisto Exclusive",
        group == "sa-sal" ~ "Salmon and SA Exclusive",
        group == "all_methods" ~ "All Methods"
      ),
      group = factor(group, levels = c("HISAT2 Exclusive",
                                       "Kallisto Exclusive",
                                       "Salmon and SA Exclusive",
                                       "All Methods"))) %>%
    ggplot(aes(x = kmer_length, y = Frequency)) +
    geom_area(fill = "royalblue",
              colour = "darkblue",
              linewidth = 1,
              alpha = 0.7) +
    labs(x = "K-mer Length (bp)",
         y = "Frequency") +
    facet_grid(~ group) +
    theme_bw() +
    theme(strip.background = element_rect(fill = "lightblue"))
}

type_nms <- new_stat_df$Type %>% unique()

for(i in type_nms) {
  kmer_stat_plotting(new_stat_df, type = i) 
  
  ggsave(filename = paste0(i, "_kmer_length_freq.png"),
         path = here("figures/kmer_analysis/dte_kmer_plots/"),
         device = "png",
         height = 150,
         width = 300,
         units = "mm",
         dpi = 400)
}


hist_df <- rbind(hisat2_dte_mer_hist,
                 kallisto_dte_mer_hist,
                 star_dte_mer_hist,
                 all_methods_dte_mer_hist)

hist_df %>%
  dplyr::filter(kmer_length >= 31) %>%
  ggplot(aes(x = kmer_length, y = Frequency)) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ group) +
  theme_bw()
```

# Kmer Analysis by DTU
## HISAT2 Alignment
```{r hisat2_dtu_kmer}
hisat2_dtu_exclusive <- hisat2_dtu_exclusive %>%
  dplyr::filter(transcript_id %in% gene_txp_anno$transcript_id)

yTx_hisat2_dtu <- exonsBy(
  edb,
  filter = TxIdFilter(hisat2_dtu_exclusive$transcript_id)
  )

yTxSeqs_hisat2_dtu <- extractTranscriptSeqs(dna, yTx_hisat2_dtu)

hisat2_dtu_kmer_df <- yTxSeqs_hisat2_dtu %>%
  as.data.frame() %>%
  rownames_to_column("transcript_id") %>%
  mutate(transcript_id = paste0(">", transcript_id)) %>%
  set_colnames(c("transcript_id", "sequence")) %>%
  as_tibble()

hisat2_dtu_kmer_df %>%
  head(100) %>%
  write_delim(here("data/kmer_analysis/loadings/hisat2_dtu/hisat2_seqs.txt"),
            delim = "\n",
            eol = "\n\n",
            col_names = FALSE)
```

```{sh eval=FALSE}
loc=/Users/konstantinosbogias/Documents/PhD/R_Projects/MethodsChapter/data/kmer_analysis

cp ${loc}/hisat2_dtu/hisat2_seqs.txt ${loc}/hisat2_dtu/hisat2_seqs.fa
for i in {1..100}; do
    jellyfish count -m ${i} -s 100M --output ${loc}/hisat2_dtu/hisat2_mer_counts.jf -C ${loc}/hisat2_dtu/hisat2_seqs.fa
    jellyfish histo ${loc}/hisat2_dtu/hisat2_mer_counts.jf > ${loc}/hisat2_dtu/histos/hisat2_${i}_mer_counts_histo.txt
    jellyfish stats ${loc}/hisat2_dtu/hisat2_mer_counts.jf > ${loc}/hisat2_dtu/stats/hisat2_${i}_mer_counts_stats.txt
    echo "hisat2 ${i}_mer done"
done
```

### Kmer data
```{r kmer_plots_hisat2_dtu eval=TRUE}
hisat2_dtu_mer_hist <- data.frame()
for(file in list.files(here("data/kmer_analysis/loadings/hisat2_dtu/histos"))) {
  hisat2_dtu_mer_add <- read_delim(
    paste0(here("data/kmer_analysis/loadings/hisat2_dtu/histos/"), file),
    col_names = FALSE,
    delim = " ") %>%
    set_colnames(c("Repeats", "Frequency")) %>%
    mutate(kmer_length = file %>%
             str_extract("...mer") %>%
             str_remove(".*_") %>% 
             str_remove("mer") %>% 
             as.numeric())
  
  hisat2_dtu_mer_hist <- rbind(hisat2_dtu_mer_hist, hisat2_dtu_mer_add)
}

hisat2_dtu_mer_hist <- hisat2_dtu_mer_hist %>%
  mutate(group = "hisat2")

write_csv(hisat2_dtu_mer_hist,
          here("data/kmer_analysis/loadings/hisat2_dtu/hisat2_mer_hist.csv"))

hisat2_dtu_mer_stat <- data.frame()
for(file in list.files(here("data/kmer_analysis/loadings/hisat2_dtu/stats"))) {
  hisat2_dtu_stat_add <- read_delim(
    paste0(here("data/kmer_analysis/loadings/hisat2_dtu/stats/"), file),
    col_names = FALSE,
    delim = " ") %>% 
    as.data.frame() %>%
    column_to_rownames("X1") %>%
    mutate_if(is.character, as.numeric) %>%
    mutate(Frequency = coalesce(X2, X3, X4, X5)) %>%
    rownames_to_column("Type") %>%
    dplyr::select("Type", "Frequency") %>%
    mutate(Type = str_remove(Type, ":")) %>% 
    dplyr::mutate(kmer_length = file %>% 
                    str_extract("...mer") %>% 
                    str_remove(".*_") %>% 
                    str_remove("mer") %>%
                    as.numeric())
  
  hisat2_dtu_mer_stat <- rbind(hisat2_dtu_mer_stat, hisat2_dtu_stat_add)
}

hisat2_dtu_mer_stat <- hisat2_dtu_mer_stat %>%
  mutate(group = "hisat2")

write_csv(hisat2_dtu_mer_stat,
          here("data/kmer_analysis/loadings/hisat2_dtu/hisat2_mer_stat.csv"))
```

### Kmer plotting
Wanna try and find long genes with less distinct kmers, so low Distinct and Unique, with a high Total. Possibly a low Max count but a high Repeat number.
```{r}
hisat2_dtu_lengths <- txp_gene_ensdb_lengths %>%
  dplyr::filter(tx_id %in% kallisto_loadings$transcript_id) %>%
  dplyr::select(tx_id, transcript_length, seqnames, strand, tx_biotype) %>%
  dplyr::mutate(group = "kallisto")
hisat2_dtu_lengths$transcript_length %>% summary()

# Summarise kmer lengths
hisat2_dtu_mer_hist$kmer_length %>% table() %>% as.data.frame()

hisat2_dtu_mer_hist %>%
  dplyr::filter(kmer_length >= 31) %>%
  ggplot(aes(x = kmer_length, y = Repeats)) + 
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7)

hisat2_dtu_mer_hist %>%
  dplyr::filter(kmer_length >= 31) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ as.character(Repeats),
             scales = "free") +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"))

hisat2_dtu_mer_hist %>%
  dplyr::filter(kmer_length <= 20) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"),
        legend.position = "none")

hisat2_dtu_mer_stat %>%
  dplyr::filter(!Type == "Max_count") %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = Type)) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ Type) +
  theme_bw()
```

## Kallisto Alignment
```{r kallisto_dtu_kmer}
kallisto_dtu_exclusive <- kallisto_dtu_exclusive %>%
  dplyr::filter(transcript_id %in% gene_txp_anno$transcript_id)

yTx_kallisto_dtu <- exonsBy(edb, filter = TxIdFilter(kallisto_dtu_exclusive$transcript_id))

yTxSeqs_kallisto_dtu <- extractTranscriptSeqs(dna, yTx_kallisto_dtu)

kallisto_dtu_kmer_df <- yTxSeqs_kallisto_dtu %>%
  as.data.frame() %>%
  rownames_to_column("transcript_id") %>%
  mutate(transcript_id = paste0(">", transcript_id)) %>%
  set_colnames(c("transcript_id", "sequence")) %>%
  as_tibble()

kallisto_dtu_kmer_df %>%
  head(100) %>%
  write_delim(here("data/kmer_analysis/loadings/kallisto_dtu/kallisto_seqs.txt"),
            delim = "\n",
            eol = "\n\n",
            col_names = FALSE)
```

```{sh eval=FALSE}
loc=/Users/konstantinosbogias/Documents/PhD/R_Projects/MethodsChapter/data/kmer_analysis

cp ${loc}/kallisto_dtu/kallisto_seqs.txt ${loc}/kallisto_dtu/kallisto_seqs.fa
for i in {1..100}; do
    jellyfish count -m ${i} -s 100M --output ${loc}/kallisto_dtu/kallisto_mer_counts.jf -C ${loc}/kallisto_dtu/kallisto_seqs.fa
    jellyfish histo ${loc}/kallisto_dtu/kallisto_mer_counts.jf > ${loc}/kallisto_dtu/histos/kallisto_${i}_mer_counts_histo.txt
    jellyfish stats ${loc}/kallisto_dtu/kallisto_mer_counts.jf > ${loc}/kallisto_dtu/stats/kallisto_${i}_mer_counts_stats.txt
    echo "kallisto ${i}_mer done"
done
```

### Kmer data
```{r kmer_plots_kallisto_dtu eval=TRUE}
kallisto_dtu_mer_hist <- data.frame()
for(file in list.files(here("data/kmer_analysis/loadings/kallisto_dtu/histos"))) {
  kallisto_dtu_mer_add <- read_delim(
    paste0(here("data/kmer_analysis/loadings/kallisto_dtu/histos/"), file),
    col_names = FALSE,
    delim = " ") %>%
    set_colnames(c("Repeats", "Frequency")) %>%
    mutate(kmer_length = file %>%
             str_extract("...mer") %>%
             str_remove(".*_") %>% 
             str_remove("mer") %>% 
             as.numeric())
  
  kallisto_dtu_mer_hist <- rbind(kallisto_dtu_mer_hist, kallisto_dtu_mer_add)
}

kallisto_dtu_mer_hist <- kallisto_dtu_mer_hist %>%
  mutate(group = "kallisto")

write_csv(kallisto_dtu_mer_hist,
          here("data/kmer_analysis/loadings/kallisto_dtu/kallisto_mer_hist.csv"))

kallisto_dtu_mer_stat <- data.frame()
for(file in list.files(here("data/kmer_analysis/loadings/kallisto_dtu/stats"))) {
  kallisto_dtu_stat_add <- read_delim(
    paste0(here("data/kmer_analysis/loadings/kallisto_dtu/stats/"), file),
    col_names = FALSE,
    delim = " ") %>% 
    as.data.frame() %>%
    column_to_rownames("X1") %>%
    mutate_if(is.character, as.numeric) %>%
    mutate(Frequency = coalesce(X2, X3, X4, X5)) %>%
    rownames_to_column("Type") %>%
    dplyr::select("Type", "Frequency") %>%
    mutate(Type = str_remove(Type, ":")) %>% 
    dplyr::mutate(kmer_length = file %>% 
                    str_extract("...mer") %>% 
                    str_remove(".*_") %>% 
                    str_remove("mer") %>%
                    as.numeric())
  
  kallisto_dtu_mer_stat <- rbind(kallisto_dtu_mer_stat, kallisto_dtu_stat_add)
}

kallisto_dtu_mer_stat <- kallisto_dtu_mer_stat %>%
  mutate(group = "kallisto")

write_csv(kallisto_dtu_mer_stat,
          here("data/kmer_analysis/loadings/kallisto_dtu/kallisto_mer_stat.csv"))
```

### Kmer plotting
Wanna try and find long genes with less distinct kmers, so low Distinct and Unique, with a high Total. Possibly a low Max count but a high Repeat number.
```{r}
kallisto_dtu_lengths <- txp_gene_ensdb_lengths %>%
  dplyr::filter(tx_id %in% kallisto_loadings$transcript_id) %>%
  dplyr::select(tx_id, transcript_length, seqnames, strand, tx_biotype) %>%
  dplyr::mutate(group = "kallisto")
kallisto_dtu_lengths$transcript_length %>% summary()

# Summarise kmer lengths
kallisto_dtu_mer_hist$kmer_length %>% table() %>% as.data.frame()

kallisto_dtu_mer_hist %>%
  dplyr::filter(kmer_length >= 31) %>%
  ggplot(aes(x = kmer_length, y = Repeats)) + 
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7)

kallisto_dtu_mer_hist %>%
  dplyr::filter(kmer_length >= 31) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ as.character(Repeats),
             scales = "free") +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"))

kallisto_dtu_mer_hist %>%
  dplyr::filter(kmer_length <= 20) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"),
        legend.position = "none")

kallisto_dtu_mer_stat %>%
  dplyr::filter(!Type == "Max_count") %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = Type)) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ Type) +
  theme_bw()
```

## STAR
```{r star_kmer}
star_dtu_exclusive <- star_dtu_exclusive %>%
  dplyr::filter(transcript_id %in% gene_txp_anno$transcript_id)

yTx_star_dtu <- exonsBy(edb, filter = TxIdFilter(star_dtu_exclusive$transcript_id))

yTxSeqs_star_dtu <- extractTranscriptSeqs(dna, yTx_star_dtu)

star_dtu_kmer_df <- yTxSeqs_star_dtu %>%
  as.data.frame() %>%
  rownames_to_column("transcript_id") %>%
  mutate(transcript_id = paste0(">", transcript_id)) %>%
  set_colnames(c("transcript_id", "sequence")) %>%
  as_tibble()

star_dtu_kmer_df %>%
  head(100) %>%
  write_delim(here("data/kmer_analysis/loadings/star_dtu/star_seqs.txt"),
            delim = "\n",
            eol = "\n\n",
            col_names = FALSE)
```

```{sh eval=FALSE}
loc=/Users/konstantinosbogias/Documents/PhD/R_Projects/MethodsChapter/data/kmer_analysis

cp ${loc}/star_dtu/star_seqs.txt ${loc}/star_dtu/star_seqs.fa
for i in {1..100}; do
    jellyfish count -m ${i} -s 100M --output ${loc}/star_dtu/star_mer_counts.jf -C ${loc}/star_dtu/star_seqs.fa
    jellyfish histo ${loc}/star_dtu/star_mer_counts.jf > ${loc}/star_dtu/histos/star_${i}_mer_counts_histo.txt
    jellyfish stats ${loc}/star_dtu/star_mer_counts.jf > ${loc}/star_dtu/stats/star_${i}_mer_counts_stats.txt
    echo "sa ${i}_mer done"
done
```

### Kmer data
```{r kmer_plots_sa eval=TRUE}
star_dtu_mer_hist <- data.frame()
for(file in list.files(here("data/kmer_analysis/loadings/star_dtu/histos"))) {
  star_dtu_mer_add <- read_delim(
    paste0(here("data/kmer_analysis/loadings/star_dtu/histos/"), file),
    col_names = FALSE,
    delim = " ") %>%
    set_colnames(c("Repeats", "Frequency")) %>%
    mutate(kmer_length = file %>%
             str_extract("...mer") %>%
             str_remove(".*_") %>% 
             str_remove("mer") %>% 
             as.numeric())
  
  star_dtu_mer_hist <- rbind(star_dtu_mer_hist, star_dtu_mer_add)
}

star_dtu_mer_hist <- star_dtu_mer_hist %>%
  mutate(group = "star")

write_csv(star_dtu_mer_hist,
          here("data/kmer_analysis/loadings/star_dtu/star_mer_hist.csv"))

star_dtu_mer_stat <- data.frame()
for(file in list.files(here("data/kmer_analysis/loadings/star_dtu/stats"))) {
  star_dtu_stat_add <- read_delim(
    paste0(here("data/kmer_analysis/loadings/star_dtu/stats/"), file),
    col_names = FALSE,
    delim = " ") %>% 
    as.data.frame() %>%
    column_to_rownames("X1") %>%
    mutate_if(is.character, as.numeric) %>%
    mutate(Frequency = coalesce(X2, X3, X4, X5)) %>%
    rownames_to_column("Type") %>%
    dplyr::select("Type", "Frequency") %>%
    mutate(Type = str_remove(Type, ":")) %>% 
    dplyr::mutate(kmer_length = file %>% 
                    str_extract("...mer") %>% 
                    str_remove(".*_") %>% 
                    str_remove("mer") %>%
                    as.numeric())
  
  star_dtu_mer_stat <- rbind(star_dtu_mer_stat, star_dtu_stat_add)
}

star_dtu_mer_stat <- star_dtu_mer_stat %>%
  mutate(group = "star")

write_csv(star_dtu_mer_stat,
          here("data/kmer_analysis/loadings/star_dtu/star_mer_stat.csv"))
```

### Kmer plotting
Wanna try and find long genes with less distinct kmers, so low Distinct and Unique, with a high Total. Possibly a low Max count but a high Repeat number.
```{r}
star_dtu_lengths <- txp_gene_ensdb_lengths %>%
  dplyr::filter(tx_id %in% star_loadings$transcript_id) %>%
  dplyr::select(tx_id, transcript_length, seqnames, strand, tx_biotype) %>%
  dplyr::mutate(group = "star")
star_dtu_lengths$transcript_length %>% summary()

# Summarise kmer lengths
star_dtu_mer_hist$kmer_length %>% table() %>% as.data.frame()

star_dtu_mer_hist %>%
  dplyr::filter(kmer_length >= 31) %>%
  ggplot(aes(x = kmer_length, y = Repeats)) + 
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7)

star_dtu_mer_hist %>%
  dplyr::filter(kmer_length >= 31) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ as.character(Repeats),
             scales = "free") +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"))

star_dtu_mer_hist %>%
  dplyr::filter(kmer_length <= 20) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"),
        legend.position = "none")

star_dtu_mer_stat %>%
  dplyr::filter(!Type == "Max_count") %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = Type)) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ Type) +
  theme_bw()
```

## Salmon
```{r star_kmer}
salmon_dtu_exclusive <- salmon_dtu_exclusive %>%
  dplyr::filter(transcript_id %in% gene_txp_anno$transcript_id)

yTx_salmon_dtu <- exonsBy(edb, filter = TxIdFilter(salmon_dtu_exclusive$transcript_id))

yTxSeqs_salmon_dtu <- extractTranscriptSeqs(dna, yTx_salmon_dtu)

salmon_dtu_kmer_df <- yTxSeqs_salmon_dtu %>%
  as.data.frame() %>%
  rownames_to_column("transcript_id") %>%
  mutate(transcript_id = paste0(">", transcript_id)) %>%
  set_colnames(c("transcript_id", "sequence")) %>%
  as_tibble()

salmon_dtu_kmer_df %>%
  head(100) %>%
  write_delim(here("data/kmer_analysis/loadings/salmon_dtu/salmon_seqs.txt"),
            delim = "\n",
            eol = "\n\n",
            col_names = FALSE)
```

```{sh eval=FALSE}
loc=/Users/konstantinosbogias/Documents/PhD/R_Projects/MethodsChapter/data/kmer_analysis

cp ${loc}/salmon_dtu/salmon_seqs.txt ${loc}/salmon_dtu/salmon_seqs.fa
for i in {1..100}; do
    jellyfish count -m ${i} -s 100M --output ${loc}/salmon_dtu/salmon_mer_counts.jf -C ${loc}/salmon_dtu/salmon_seqs.fa
    jellyfish histo ${loc}/salmon_dtu/salmon_mer_counts.jf > ${loc}/salmon_dtu/histos/salmon_${i}_mer_counts_histo.txt
    jellyfish stats ${loc}/salmon_dtu/salmon_mer_counts.jf > ${loc}/salmon_dtu/stats/salmon_${i}_mer_counts_stats.txt
    echo "salmon ${i}_mer done"
done
```

### Kmer data
```{r kmer_plots_sa eval=TRUE}
salmon_dtu_mer_hist <- data.frame()
for(file in list.files(here("data/kmer_analysis/loadings/salmon_dtu/histos"))) {
  salmon_dtu_mer_add <- read_delim(
    paste0(here("data/kmer_analysis/loadings/salmon_dtu/histos/"), file),
    col_names = FALSE,
    delim = " ") %>%
    set_colnames(c("Repeats", "Frequency")) %>%
    mutate(kmer_length = file %>%
             str_extract("...mer") %>%
             str_remove(".*_") %>% 
             str_remove("mer") %>% 
             as.numeric())
  
  salmon_dtu_mer_hist <- rbind(salmon_dtu_mer_hist, salmon_dtu_mer_add)
}

salmon_dtu_mer_hist <- salmon_dtu_mer_hist %>%
  mutate(group = "salmon")

write_csv(salmon_dtu_mer_hist,
          here("data/kmer_analysis/loadings/salmon_dtu/salmon_mer_hist.csv"))

salmon_dtu_mer_stat <- data.frame()
for(file in list.files(here("data/kmer_analysis/loadings/salmon_dtu/stats"))) {
  salmon_dtu_stat_add <- read_delim(
    paste0(here("data/kmer_analysis/loadings/salmon_dtu/stats/"), file),
    col_names = FALSE,
    delim = " ") %>% 
    as.data.frame() %>%
    column_to_rownames("X1") %>%
    mutate_if(is.character, as.numeric) %>%
    mutate(Frequency = coalesce(X2, X3, X4, X5)) %>%
    rownames_to_column("Type") %>%
    dplyr::select("Type", "Frequency") %>%
    mutate(Type = str_remove(Type, ":")) %>% 
    dplyr::mutate(kmer_length = file %>% 
                    str_extract("...mer") %>% 
                    str_remove(".*_") %>% 
                    str_remove("mer") %>%
                    as.numeric())
  
  salmon_dtu_mer_stat <- rbind(salmon_dtu_mer_stat, salmon_dtu_stat_add)
}

salmon_dtu_mer_stat <- salmon_dtu_mer_stat %>%
  mutate(group = "salmon")

write_csv(salmon_dtu_mer_stat,
          here("data/kmer_analysis/loadings/salmon_dtu/salmon_mer_stat.csv"))
```

### Kmer plotting
Wanna try and find long genes with less distinct kmers, so low Distinct and Unique, with a high Total. Possibly a low Max count but a high Repeat number.
```{r}
salmon_dtu_lengths <- txp_gene_ensdb_lengths %>%
  dplyr::filter(tx_id %in% salmon_loadings$transcript_id) %>%
  dplyr::select(tx_id, transcript_length, seqnames, strand, tx_biotype) %>%
  dplyr::mutate(group = "salmon")
salmon_dtu_lengths$transcript_length %>% summary()

# Summarise kmer lengths
salmon_dtu_mer_hist$kmer_length %>% table() %>% as.data.frame()

salmon_dtu_mer_hist %>%
  dplyr::filter(kmer_length >= 31) %>%
  ggplot(aes(x = kmer_length, y = Repeats)) + 
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7)

salmon_dtu_mer_hist %>%
  dplyr::filter(kmer_length >= 31) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ as.character(Repeats),
             scales = "free") +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"))

salmon_dtu_mer_hist %>%
  dplyr::filter(kmer_length <= 20) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"),
        legend.position = "none")

salmon_dtu_mer_stat %>%
  dplyr::filter(!Type == "Max_count") %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = Type)) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ Type) +
  theme_bw()
```

## All Methods DTU
```{r star_kmer}
all_methods_dtu_tx_remove <- txp_gene_ensdb_lengths %>%
  dplyr::filter(tx_id %in% all_methods_tx_dtu$transcript_id) %>%
  dplyr::filter(str_detect(seqnames, "^CHR")) %>%
  as.data.frame()

all_methods_tx_dtu <- all_methods_tx_dtu %>%
  dplyr::filter(!transcript_id %in% all_methods_dtu_tx_remove$tx_id)

yTx_all_methods_dtu <- exonsBy(
  edb,
  filter = TxIdFilter(all_methods_tx_dtu$transcript_id)
  )

yTxSeqs_all_methods_dtu <- extractTranscriptSeqs(dna, yTx_all_methods_dtu)

all_methods_dtu_kmer_df <- yTxSeqs_all_methods_dtu %>%
  as.data.frame() %>%
  rownames_to_column("transcript_id") %>%
  mutate(transcript_id = paste0(">", transcript_id)) %>%
  set_colnames(c("transcript_id", "sequence")) %>%
  as_tibble()

all_methods_dtu_kmer_df %>%
  head(19) %>%
  write_delim(here("data/kmer_analysis/all_methods_dtu/all_methods_seqs.txt"),
              delim = "\n",
              eol = "\n\n",
              col_names = FALSE)
```

```{sh eval=FALSE}
loc=/Users/konstantinosbogias/Documents/PhD/R_Projects/MethodsChapter/data/kmer_analysis

cp ${loc}/all_methods_dtu/all_methods_seqs.txt ${loc}/all_methods_dtu/all_methods_seqs.fa
for i in {1..100}; do
    jellyfish count -m ${i} -s 100M --output ${loc}/all_methods_dtu/all_methods_mer_counts.jf -C ${loc}/all_methods_dtu/all_methods_seqs.fa
    jellyfish histo ${loc}/all_methods_dtu/all_methods_mer_counts.jf > ${loc}/all_methods_dtu/histos/all_methods_${i}_mer_counts_histo.txt
    jellyfish stats ${loc}/all_methods_dtu/all_methods_mer_counts.jf > ${loc}/all_methods_dtu/stats/all_methods_${i}_mer_counts_stats.txt
    echo "all_methods ${i}_mer done"
done
```

### Kmer data
```{r kmer_plots_salmon eval=TRUE}
all_methods_dtu_mer_hist <- data.frame()
for(file in list.files(here("data/kmer_analysis/all_methods_dtu/histos"))) {
  all_methods_dtu_mer_add <- read_delim(
    paste0(here("data/kmer_analysis/all_methods_dtu/histos/"), file),
    col_names = FALSE,
    delim = " ") %>%
    set_colnames(c("Repeats", "Frequency")) %>%
    mutate(kmer_length = file %>%
             str_extract("...mer") %>%
             str_remove(".*_") %>% 
             str_remove("mer") %>% 
             as.numeric())
  
  all_methods_dtu_mer_hist <- rbind(all_methods_dtu_mer_hist,
                                    all_methods_dtu_mer_add)
}

all_methods_dtu_mer_hist <- all_methods_dtu_mer_hist %>%
  mutate(group = "all_methods")

write_csv(all_methods_dtu_mer_hist,
          here("data/kmer_analysis/all_methods_dtu/all_methods_dtu_mer_hist.csv"))

all_methods_dtu_mer_stat <- data.frame()
for(file in list.files(here("data/kmer_analysis/all_methods_dtu/stats"))) {
  all_methods_dtu_stat_add <- read_delim(
    paste0(here("data/kmer_analysis/all_methods_dtu/stats/"), file),
    col_names = FALSE,
    delim = " ") %>% 
    as.data.frame() %>%
    column_to_rownames("X1") %>%
    mutate_if(is.character, as.numeric) %>%
    mutate(Frequency = coalesce(X2, X3, X4, X5)) %>%
    rownames_to_column("Type") %>%
    dplyr::select("Type", "Frequency") %>%
    mutate(Type = str_remove(Type, ":")) %>% 
    dplyr::mutate(kmer_length = file %>% 
                    str_extract("...mer") %>% 
                    str_remove(".*_") %>% 
                    str_remove("mer") %>%
                    as.numeric())
  
  all_methods_dtu_mer_stat <- rbind(all_methods_dtu_mer_stat,
                                    all_methods_dtu_stat_add)
}

all_methods_dtu_mer_stat <- all_methods_dtu_mer_stat %>%
  mutate(group = "all_methods")

write_csv(all_methods_dtu_mer_stat,
          here("data/kmer_analysis/all_methods_dtu/all_methods_dtu_mer_stat.csv"))
```

### Kmer plotting
Wanna try and find long genes with less distinct kmers, so low Distinct and Unique, with a high Total. Possibly a low Max count but a high Repeat number.
```{r}
all_methods_dtu_lengths <- txp_gene_ensdb_lengths %>%
  dplyr::filter(tx_id %in% all_methods_tx_dtu$transcript_id) %>%
  dplyr::select(tx_id, transcript_length, seqnames, strand, tx_biotype) %>%
  dplyr::mutate(group = "all_methods")
all_methods_dtu_lengths$transcript_length %>% summary()

# Summarise kmer lengths
all_methods_dtu_mer_hist$kmer_length %>% table() %>% as.data.frame()

all_methods_dtu_mer_hist %>%
  dplyr::filter(kmer_length >= 31) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ as.character(Repeats),
             scales = "free") +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"))

all_methods_dtu_mer_hist %>%
  dplyr::filter(kmer_length <= 20) %>%
  dplyr::mutate(Repeats = paste0("Repeat ", as.character(Repeats))) %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = as.character(Repeats))) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  labs(x = "K-mer Length (bp)",
       y = "Frequency",
       fill = "Kmer") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 10),
        axis.title = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        strip.background = element_rect(fill = "lightblue"),
        legend.position = "none")

all_methods_dtu_mer_stat %>%
  dplyr::filter(!Type == "Max_count") %>%
  ggplot(aes(x = kmer_length, y = Frequency, fill = Type)) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ Type) +
  theme_bw()
```

## DTU Plotting
```{r}
lengths_df <- rbind(hisat2_dtu_lengths,
                    kallisto_dtu_lengths,
                    star_dtu_lengths,
                    salmon_dtu_lengths,
                    all_methods_dtu_lengths)

lengths_by_method <- lengths_df %>%
  dplyr::mutate(
    group = case_when(
      group == "all_methods" ~ "All Methods",
      group == "hisat2" ~ "HISAT2 Exclusive",
      group == "kallisto" ~ "Kallisto Exclusive",
      group == "star" ~ "STAR Exclusive",
      group == "salmon" ~ "Salmon Exclusive"
    ),
    group = factor(group, levels = c("All Methods",
                                     "HISAT2 Exclusive",
                                     "Kallisto Exclusive",
                                     "Selective Alignment Exclusive",
                                     "Salmon Exclusive"))) %>%
  ggplot(aes(x = group, y = transcript_length)) +
  geom_boxplot(fill = "lightblue") +
  scale_y_log10() +
  labs(x = "",
       y = "Transcript Length (log10)") +
  theme_bw()

lengths_by_method %>%
  ggsave(filename = "length_by_method_plot.png",
         path = here("figures/kmer_analysis/dtu_kmer_plots/"),
         device = "png",
         height = 200,
         width = 200,
         units = "mm",
         dpi = 400)

txp_lengths_hist <- txp_gene_ensdb_lengths %>%
  dplyr::select(tx_id, transcript_length) %>%
  mutate(group = "All Transcripts") %>%
  rbind(lengths_df %>%
          dplyr::select(tx_id, transcript_length) %>%
          dplyr::mutate(group = "Top Loadings Transcripts")) %>%
  ggplot(aes(x = log10(transcript_length))) +
  geom_histogram(bins = 200) +
  scale_x_continuous(limits = c(1, 6),
                     breaks = c(1, 2, 3, 4, 5, 6)) +
  labs(x = "Transcript Length (log10)",
       y = "Frequency") +
  facet_grid(group ~ ., switch = "y", scales = "free") +
  theme_bw() +
  theme(axis.text = element_text(colour = "black", size = 12),
        axis.title = element_text(colour = "black", size = 14),
        strip.background = element_rect(fill = "lightblue"),
        strip.text = element_text(colour = "black", size = 14))

txp_lengths_hist %>%
  ggsave(filename = "length_plot.png",
         path = here("figures/kmer_analysis/dtu_kmer_plots/"),
         device = "png",
         height = 150,
         width = 200,
         units = "mm",
         dpi = 400)

# We can see that there is stuff all variation in frequency within groups
stat_df <- rbind(all_methods_dtu_mer_stat,
                 hisat2_dtu_mer_stat,
                 kallisto_dtu_mer_stat,
                 star_dtu_mer_stat,
                 salmon_dtu_mer_stat)

group_names <- stat_df$group %>% unique()
new_stat_df <- data.frame()
stat_group_chunk <- data.frame()
for(g in group_names) {
  for(i in 1:100) {
    stat_kmer_chunk <- stat_df %>%
      dplyr::filter(kmer_length == i) %>%
      dplyr::filter(group == g) %>%
      dplyr::select(Type, Frequency) %>%
      column_to_rownames("Type") %>%
      t() %>%
      as_tibble() %>%
      mutate(Multiplicity = Total-Distinct) %>%
      t() %>%
      set_colnames("Frequency") %>%
      as.data.frame() %>%
      rownames_to_column("Type") %>%
      mutate(kmer_length = i, group = g)
    
    stat_group_chunk <- rbind(stat_group_chunk, stat_kmer_chunk)
  }
  new_stat_df <- rbind(new_stat_df, stat_group_chunk)
}


kmer_stat_plotting <- function(x, type, kmer = 31) {
  x %>%
    dplyr::filter(kmer_length >= kmer,
                  Type == type) %>%
    dplyr::mutate(
      group = case_when(
        group == "all_methods" ~ "All Methods",
        group == "hisat2" ~ "HISAT2 Exclusive",
        group == "kallisto" ~ "Kallisto Exclusive",
        group == "star" ~ "STAR Exclusive",
        group == "salmon" ~ "Salmon Exclusive"
      ),
      group = factor(group, levels = c("All Methods",
                                       "HISAT2 Exclusive",
                                       "Kallisto Exclusive",
                                       "Selective Alignment Exclusive",
                                       "Salmon Exclusive"))) %>%
    ggplot(aes(x = kmer_length, y = Frequency)) +
    geom_area(fill = "royalblue",
              colour = "darkblue",
              linewidth = 1,
              alpha = 0.7) +
    labs(x = "K-mer Length (bp)",
         y = "Frequency") +
    facet_grid(~ group) +
    theme_bw() +
    theme(strip.background = element_rect(fill = "lightblue"))
}

type_nms <- new_stat_df$Type %>% unique()

for(i in type_nms) {
  kmer_stat_plotting(new_stat_df, type = i) 
  
  ggsave(filename = paste0(i, "_kmer_length_freq.png"),
         path = here("figures/kmer_analysis/dtu_kmer_plots/"),
         device = "png",
         height = 150,
         width = 300,
         units = "mm",
         dpi = 400)
}


hist_df <- rbind(all_methods_dtu_mer_hist,
                 hisat2_dtu_mer_hist,
                 kallisto_dtu_mer_hist,
                 star_dtu_mer_hist,
                 salmon_dtu_mer_hist)

hist_df %>%
  dplyr::filter(kmer_length >= 31 & Repeats == 2) %>%
  ggplot(aes(x = kmer_length, y = Frequency)) +
  geom_area(fill = "royalblue",
            colour = "darkblue",
            linewidth = 1,
            alpha = 0.7) +
  facet_grid(~ group) +
  theme_bw()
```
